<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Abhinav Mishra" /><link rel="canonical" href="https://bibymaths.github.io/phoskintime/reference.html" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>Reference - phoskintime</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="assets/_mkdocstrings.css" rel="stylesheet" />
        <link href="styles/overrides.css" rel="stylesheet" />
        <link href="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Reference";
        var mkdocs_page_input_path = "reference.md";
        var mkdocs_page_url = "/phoskintime/reference.html";
      </script>
    
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="index.html" class="icon icon-home"> phoskintime
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="index.html">Home</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Reference</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#data-standardization-cleanup">Data Standardization &amp; Cleanup</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#processing.cleanup">cleanup</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#processing.cleanup.process_collecttri">process_collecttri</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.cleanup.format_site">format_site</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.cleanup.process_msgauss">process_msgauss</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.cleanup.process_msgauss_std">process_msgauss_std</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.cleanup.process_routlimma">process_routlimma</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.cleanup.update_gene_symbols">update_gene_symbols</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.cleanup.move_processed_files">move_processed_files</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#optimization-results-mapping">Optimization Results Mapping</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#processing.map">map</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#processing.map.map_optimization_results">map_optimization_results</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.map.create_cytoscape_table">create_cytoscape_table</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.map.add_kinetic_strength_columns">add_kinetic_strength_columns</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#processing.map.generate_nodes">generate_nodes</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinase-phosphorylation-optimization">Kinase-Phosphorylation Optimization</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#evolutionary-algorithms">Evolutionary Algorithms</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.config.constants">constants</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.config.logconf">logconf</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.config.logconf.ColoredFormatter">ColoredFormatter</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#kinopt.evol.config.logconf.ColoredFormatter.format">format</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#kinopt.evol.config.logconf.ColoredFormatter.remove_ansi">remove_ansi</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.config.logconf.setup_logger">setup_logger</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.exporter.plotout">plotout</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.exporter.plotout.plot_residuals_for_gene">plot_residuals_for_gene</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.exporter.plotout.opt_analyze_nsga">opt_analyze_nsga</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.exporter.plotout.opt_analyze_de">opt_analyze_de</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.exporter.sheetutils">sheetutils</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.exporter.sheetutils.output_results">output_results</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.objfn.minfndiffevo">minfndiffevo</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem">PhosphorylationOptimizationProblem</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__">__init__</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function">objective_function</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.objfn.minfnnsgaii">minfnnsgaii</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem">PhosphorylationOptimizationProblem</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__">__init__</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function">objective_function</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.opt.optrun">optrun</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.opt.optrun.run_optimization">run_optimization</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.opt.optrun.post_optimization_nsga">post_optimization_nsga</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.opt.optrun.post_optimization_de">post_optimization_de</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.optcon.construct">construct</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.optcon.construct.pipeline">pipeline</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.optcon.construct.load_geneid_to_psites">load_geneid_to_psites</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.optcon.construct.get_unique_kinases">get_unique_kinases</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.optcon.construct.check_kinases">check_kinases</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.utils.iodata">iodata</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.utils.iodata.format_duration">format_duration</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.utils.iodata.load_and_scale_data">load_and_scale_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.utils.iodata.apply_scaling">apply_scaling</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.utils.iodata.create_report">create_report</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.utils.iodata.organize_output_files">organize_output_files</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.evol.utils.params">params</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.utils.params.extract_parameters">extract_parameters</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.evol.utils.params.compute_metrics">compute_metrics</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#gradient-based-algorithms">Gradient-Based Algorithms</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.config.constants">constants</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.config.constants.parse_args">parse_args</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.config.logconf">logconf</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.exporter.plotout">plotout</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.plotout.plot_fits_for_gene">plot_fits_for_gene</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.plotout.plot_cumulative_residuals">plot_cumulative_residuals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.plotout.plot_autocorrelation_residuals">plot_autocorrelation_residuals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.plotout.plot_histogram_residuals">plot_histogram_residuals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.plotout.plot_qqplot_residuals">plot_qqplot_residuals</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.exporter.sheetutils">sheetutils</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.sheetutils.plot_fits_for_gene">plot_fits_for_gene</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.sheetutils.plot_cumulative_residuals">plot_cumulative_residuals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals">plot_autocorrelation_residuals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.sheetutils.plot_histogram_residuals">plot_histogram_residuals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.sheetutils.plot_qqplot_residuals">plot_qqplot_residuals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.exporter.sheetutils.output_results">output_results</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.objfn.minfn">minfn</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.opt.optrun">optrun</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.opt.optrun.run_optimization">run_optimization</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.optcon.construct">construct</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.optcon.construct.load_geneid_to_psites">load_geneid_to_psites</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.optcon.construct.get_unique_kinases">get_unique_kinases</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.optcon.construct.check_kinases">check_kinases</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.utils.iodata">iodata</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.utils.iodata.format_duration">format_duration</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.utils.iodata.load_and_scale_data">load_and_scale_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.utils.iodata.apply_scaling">apply_scaling</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.utils.iodata.create_report">create_report</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.utils.iodata.organize_output_files">organize_output_files</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.local.utils.params">params</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.utils.params.extract_parameters">extract_parameters</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.local.utils.params.compute_metrics">compute_metrics</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#fitting-analysis-feasibility">Fitting Analysis &amp; Feasibility</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit">postfit</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.goodnessoffit">goodnessoffit</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta">reshape_alpha_beta</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.perform_pca">perform_pca</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.plot_pca">plot_pca</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.perform_tsne">perform_tsne</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.additional_plots">additional_plots</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.create_sankey_from_network">create_sankey_from_network</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.fitanalysis.helpers.postfit.important_connections">important_connections</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#kinopt.optimality.KKT">KKT</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.optimality.KKT.generate_latex_table">generate_latex_table</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.optimality.KKT.print_primal_feasibility_results">print_primal_feasibility_results</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.optimality.KKT.print_sensitivity_and_active_constraints">print_sensitivity_and_active_constraints</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.optimality.KKT.plot_constraint_violations">plot_constraint_violations</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.optimality.KKT.plot_sensitivity_analysis">plot_sensitivity_analysis</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.optimality.KKT.process_excel_results">process_excel_results</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#kinopt.optimality.KKT.post_optimization_results">post_optimization_results</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tf-mrna-optimization">TF-mRNA Optimization</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#evolutionary-algorithms_1">Evolutionary Algorithms</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.config.constants">constants</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.config.constants.parse_args">parse_args</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.config.logconf">logconf</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.exporter.plotout">plotout</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.exporter.plotout.plot_estimated_vs_observed">plot_estimated_vs_observed</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.exporter.plotout.compute_predictions">compute_predictions</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.exporter.sheetutils">sheetutils</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.exporter.sheetutils.save_results_to_excel">save_results_to_excel</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.objfn.minfn">minfn</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem">TFOptimizationMultiObjectiveProblem</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__">__init__</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.objfn.minfn.objective_">objective_</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.opt.optrun">optrun</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.opt.optrun.run_optimization">run_optimization</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.optcon.construct">construct</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.optcon.construct.build_fixed_arrays">build_fixed_arrays</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.optcon.filter">filter</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.optcon.filter.load_raw_data">load_raw_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.optcon.filter.filter_mrna">filter_mrna</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.optcon.filter.update_regulations">update_regulations</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.optcon.filter.filter_TF">filter_TF</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.optcon.filter.determine_T_use">determine_T_use</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.utils.iodata">iodata</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.iodata.load_mRNA_data">load_mRNA_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.iodata.load_TF_data">load_TF_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.iodata.load_regulation">load_regulation</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.iodata.create_report">create_report</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.iodata.organize_output_files">organize_output_files</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.iodata.format_duration">format_duration</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.evol.utils.params">params</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.create_no_psite_array">create_no_psite_array</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.compute_beta_indices">compute_beta_indices</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.create_initial_guess">create_initial_guess</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.create_bounds">create_bounds</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.get_parallel_runner">get_parallel_runner</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.extract_best_solution">extract_best_solution</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.print_alpha_mapping">print_alpha_mapping</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.evol.utils.params.print_beta_mapping">print_beta_mapping</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#gradient-based-algorithms_1">Gradient-Based Algorithms</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.config.constants">constants</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.config.constants.parse_args">parse_args</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.config.logconf">logconf</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.exporter.plotout">plotout</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.exporter.plotout.plot_estimated_vs_observed">plot_estimated_vs_observed</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.exporter.sheetutils">sheetutils</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.exporter.sheetutils.save_results_to_excel">save_results_to_excel</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.objfn.minfn">minfn</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.objfn.minfn.objective_">objective_</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.objfn.minfn.compute_predictions">compute_predictions</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.objfn.minfn.objective_wrapper">objective_wrapper</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.opt.optrun">optrun</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.opt.optrun.run_optimizer">run_optimizer</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.optcon.construct">construct</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.optcon.construct.build_fixed_arrays">build_fixed_arrays</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.optcon.construct.constraint_alpha_func">constraint_alpha_func</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.optcon.construct.constraint_beta_func">constraint_beta_func</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.optcon.construct.build_linear_constraints">build_linear_constraints</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.optcon.filter">filter</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.optcon.filter.load_and_filter_data">load_and_filter_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.optcon.filter.prepare_data">prepare_data</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.utils.iodata">iodata</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.iodata.min_max_normalize">min_max_normalize</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.iodata.load_expression_data">load_expression_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.iodata.load_tf_protein_data">load_tf_protein_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.iodata.load_regulation">load_regulation</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.iodata.summarize_stats">summarize_stats</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.iodata.create_report">create_report</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.iodata.organize_output_files">organize_output_files</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.local.utils.params">params</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.params.get_optimization_parameters">get_optimization_parameters</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.local.utils.params.postprocess_results">postprocess_results</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#fitting-analysis">Fitting Analysis</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#tfopt.fitanalysis.helper">helper</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter">Plotter</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.__init__">__init__</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.load_data">load_data</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution">plot_alpha_distribution</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_beta_barplots">plot_beta_barplots</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals">plot_heatmap_abs_residuals</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit">plot_goodness_of_fit</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_kld">plot_kld</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_pca">plot_pca</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha">plot_boxplot_alpha</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta">plot_boxplot_beta</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha">plot_cdf_alpha</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_cdf_beta">plot_cdf_beta</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals">plot_time_wise_residuals</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#ode-modelling-parameter-estimation">ODE Modelling &amp; Parameter Estimation</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#configuration">Configuration</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#config.cli">cli</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--usage">Usage</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--come-one-level-up-from-the-package-root-it-should-be-the-working-directory">Come one level up from the package root, it should be the working directory</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--where-you-can-see-the-project-directory">(where you can see the project directory).</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--run-everything-with-the-default-local-solver">run everything with the default (local) solver</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--run-only-preprocessing">run only preprocessing</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--run-tfopt-with-local-flavour">run tfopt with local flavour</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--run-tfopt-with-evol-flavour">run tfopt with evol flavour</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--run-kinopt-with-local-flavour">run kinopt with local flavour</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--run-kinopt-with-evol-flavour">run kinopt with evol flavour</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli--run-the-model">run the model</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli.prep">prep</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli.tfopt">tfopt</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli.kinopt">kinopt</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli.model">model</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.cli.all">all</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#config.config">config</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.parse_bound_pair">parse_bound_pair</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.parse_fix_value">parse_fix_value</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.ensure_output_directory">ensure_output_directory</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.parse_args">parse_args</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.log_config">log_config</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.extract_config">extract_config</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.score_fit">score_fit</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.config.future_times">future_times</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#config.constants">constants</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#config.constants.get_param_names_rand">get_param_names_rand</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.constants.get_param_names_ds">get_param_names_ds</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.constants.generate_labels_rand">generate_labels_rand</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.constants.generate_labels_ds">generate_labels_ds</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.constants.location">location</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.constants.get_number_of_params_rand">get_number_of_params_rand</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.constants.get_bounds_rand">get_bounds_rand</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#config.logconf">logconf</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#config.logconf.ColoredFormatter">ColoredFormatter</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#config.logconf.ColoredFormatter.format">format</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#config.logconf.ColoredFormatter.remove_ansi">remove_ansi</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#config.logconf.setup_logger">setup_logger</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#core-functions">Core Functions</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#paramest.normest">normest</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#paramest.normest.worker_find_lambda">worker_find_lambda</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#paramest.normest.find_best_lambda">find_best_lambda</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#paramest.normest.normest">normest</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#paramest.toggle">toggle</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#paramest.toggle.estimate_parameters">estimate_parameters</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#weights-for-curve-fitting">Weights for Curve Fitting</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#models.weights">weights</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#models.weights.early_emphasis">early_emphasis</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.weights.get_protein_weights">get_protein_weights</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.weights.full_weight">full_weight</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.weights.get_weight_options">get_weight_options</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#parameter-estimation">Parameter Estimation</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#paramest.core">core</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#paramest.core.process_gene">process_gene</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#paramest.core.process_gene_wrapper">process_gene_wrapper</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#confidence-intervals-using-linearization">Confidence Intervals using Linearization</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#paramest.identifiability.ci">ci</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#paramest.identifiability.ci.confidence_intervals">confidence_intervals</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#knockout-analysis">Knockout Analysis</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#knockout.helper">helper</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#perturbation-parameter-sensitivity-analysis">Perturbation &amp; Parameter Sensitivity Analysis</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#sensitivity.analysis">analysis</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#sensitivity.analysis.compute_bound">compute_bound</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#sensitivity.analysis.define_sensitivity_problem_rand">define_sensitivity_problem_rand</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#sensitivity.analysis.define_sensitivity_problem_ds">define_sensitivity_problem_ds</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#model-diagram">Model Diagram</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#models.diagram.helpers">helpers</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#models.diagram.helpers.powerset">powerset</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.diagram.helpers.state_label">state_label</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.diagram.helpers.create_random_diagram">create_random_diagram</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.diagram.helpers.create_distributive_diagram">create_distributive_diagram</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.diagram.helpers.create_successive_model">create_successive_model</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#model-types">Model Types</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#models.distmod">distmod</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#models.distmod.ode_core">ode_core</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.distmod.unpack_params">unpack_params</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.distmod.solve_ode">solve_ode</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#models.randmod">randmod</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#models.randmod.unpack_params">unpack_params</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.randmod.ode_system">ode_system</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.randmod.solve_ode">solve_ode</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#models.succmod">succmod</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#models.succmod.ode_core">ode_core</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.succmod.unpack_params">unpack_params</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#models.succmod.solve_ode">solve_ode</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#steady-state-calculation">Steady-State Calculation</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#steady.initdist">initdist</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#steady.initdist.initial_condition">initial_condition</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#steady.initrand">initrand</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#steady.initrand.initial_condition">initial_condition</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#steady.initsucc">initsucc</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#steady.initsucc.initial_condition">initial_condition</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#plotting">Plotting</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#plotting.plotting">plotting</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#plotting.plotting.Plotter">Plotter</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_parallel">plot_parallel</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.pca_components">pca_components</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_pca">plot_pca</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_tsne">plot_tsne</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_param_series">plot_param_series</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_profiles">plot_profiles</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_model_fit">plot_model_fit</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_param_scatter">plot_param_scatter</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_heatmap">plot_heatmap</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_error_distribution">plot_error_distribution</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_gof">plot_gof</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_kld">plot_kld</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_params_bar">plot_params_bar</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_knockouts">plot_knockouts</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_top_param_pairs">plot_top_param_pairs</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_model_perturbations">plot_model_perturbations</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_time_state_grid">plot_time_state_grid</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_phase_space">plot_phase_space</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_future_fit">plot_future_fit</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_regularization">plot_regularization</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#plotting.plotting.Plotter.plot_model_error">plot_model_error</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utility-functions">Utility Functions</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#utils.display">display</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#utils.display.ensure_output_directory">ensure_output_directory</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.display.load_data">load_data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.display.format_duration">format_duration</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.display.merge_obs_est">merge_obs_est</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.display.save_result">save_result</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.display.create_report">create_report</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.display.organize_output_files">organize_output_files</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#utils.tables">tables</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#utils.tables.generate_tables">generate_tables</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.tables.save_tables">save_tables</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.tables.save_master_table">save_master_table</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#utils.latexit">latexit</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#utils.latexit.generate_latex_table">generate_latex_table</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.latexit.generate_latex_image">generate_latex_image</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.latexit.main">main</a>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="Documentation/CHANGELOG.html">Changelog</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/processing/index.html">Processing & Mapping</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/config/index.html">Configuration</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/kinopt/index.html">Kinase Optimization</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/tfopt/index.html">Transcription Factor Optimization</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/models/index.html">Models</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/steady/index.html">Steady State Calculation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/paramest/index.html">Parameter Estimation</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/identifiability/index.html">Confidence Intervals</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/sensitivity/index.html">Sensitivity Analysis</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/knockout/index.html">Knockout Analysis</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/plotting/index.html">Plotting</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="Documentation/utils/index.html">Utils</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">phoskintime</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html" class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Reference</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/bibymaths/phoskintime/edit/master/docs/reference.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="api-reference">API Reference</h1>
<h2 id="data-standardization-cleanup">Data Standardization &amp; Cleanup</h2>


<div class="doc doc-object doc-module">



<h2 id="processing.cleanup" class="doc doc-heading">
            <code>processing.cleanup</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_collecttri" class="doc doc-heading">
            <code class="highlight language-python">process_collecttri()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Processes the CollecTRI file to clean and filter mRNA-TF interactions.
Removes complex interactions, filters by target genes, and saves the result.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.format_site" class="doc doc-heading">
            <code class="highlight language-python">format_site(site)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Formats a phosphorylation site string.</p>
<p>If the input is NaN or an empty string, returns an empty string.
If the input contains an underscore ('_'), splits the string into two parts,
converts the first part to uppercase, and appends the second part unchanged.
Otherwise, converts the entire string to uppercase.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>site</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The phosphorylation site string to format.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>str</code></b>              –
              <div class="doc-md-description">
                <p>The formatted phosphorylation site string.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_msgauss" class="doc doc-heading">
            <code class="highlight language-python">process_msgauss()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Processes the MS Gaussian data file to generate time series data.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_msgauss_std" class="doc doc-heading">
            <code class="highlight language-python">process_msgauss_std()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Processes the MS Gaussian data file to compute transformed means and standard deviations.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.process_routlimma" class="doc doc-heading">
            <code class="highlight language-python">process_routlimma()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Processes the Rout Limma table to generate time series data for mRNA.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.update_gene_symbols" class="doc doc-heading">
            <code class="highlight language-python">update_gene_symbols(filename)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Updates the GeneID column in a CSV file by mapping GeneIDs to gene/protein symbols.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The path to the CSV file to be updated. The file must contain a 'GeneID' column.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.cleanup.move_processed_files" class="doc doc-heading">
            <code class="highlight language-python">move_processed_files()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Moves or copies processed files to their respective directories.</p>


    </div>

</div>



  </div>

    </div>

</div><h2 id="optimization-results-mapping">Optimization Results Mapping</h2>


<div class="doc doc-object doc-module">



<h2 id="processing.map" class="doc doc-heading">
            <code>processing.map</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="processing.map.map_optimization_results" class="doc doc-heading">
            <code class="highlight language-python">map_optimization_results(tf_file_path, kin_file_path, sheet_name='Alpha Values')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Reads the TF-mRNA optimization results from an Excel file and maps mRNA to each TF.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>tf_file_path</code></b>
              –
              <div class="doc-md-description">
                <p>Path to the Excel file containing TF-mRNA optimization results.</p>
              </div>
            </li>
            <li>
              <b><code>kin_file_path</code></b>
              –
              <div class="doc-md-description">
                <p>Path to the Excel file containing Kinase-Phosphorylation optimization results.</p>
              </div>
            </li>
            <li>
              <b><code>sheet_name</code></b>
              –
              <div class="doc-md-description">
                <p>The name of the sheet in the Excel file to read from. Default is 'Alpha Values'.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the mapped TF, mRNA, Psite, and Kinase information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.map.create_cytoscape_table" class="doc doc-heading">
            <code class="highlight language-python">create_cytoscape_table(mapping_csv_path)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a Cytoscape-compatible edge table from a mapping file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mapping_csv_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the input CSV file with columns:
                    TF, TF_strength, mRNA, Psite, Kinase, Kinase_strength</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>table</code></b>(                  <code><span title="pandas.DataFrame">DataFrame</span></code>
)              –
              <div class="doc-md-description">
                <p>Edge table with columns [Source, Target, Interaction, Strength]</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.map.add_kinetic_strength_columns" class="doc doc-heading">
            <code class="highlight language-python">add_kinetic_strength_columns(mapping_path, mapping__path, excel_path, suffix)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Adds kinetic strength columns to the mapping files based on the provided Excel file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mapping_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the first mapping file.</p>
              </div>
            </li>
            <li>
              <b><code>mapping__path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the second mapping file.</p>
              </div>
            </li>
            <li>
              <b><code>excel_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the Excel file containing kinetic strength data.</p>
              </div>
            </li>
            <li>
              <b><code>suffix</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Suffix to append to the output files.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="processing.map.generate_nodes" class="doc doc-heading">
            <code class="highlight language-python">generate_nodes(edge_df)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Infers node types and aggregates all phosphorylation sites per target node from phosphorylation edges.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>edge_df</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>Must have columns ['Source', 'Target', 'Interaction', 'Psite']</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: DataFrame with columns ['Node', 'Type', 'Psite']</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h2 id="kinase-phosphorylation-optimization">Kinase-Phosphorylation Optimization</h2>
<h3 id="evolutionary-algorithms">Evolutionary Algorithms</h3>


<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.config.constants" class="doc doc-heading">
            <code>kinopt.evol.config.constants</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.config.logconf" class="doc doc-heading">
            <code>kinopt.evol.config.logconf</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="kinopt.evol.config.logconf.ColoredFormatter" class="doc doc-heading">
            <code>ColoredFormatter</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="logging.Formatter">Formatter</span></code></p>











  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.config.logconf.ColoredFormatter.format" class="doc doc-heading">
            <code class="highlight language-python">format(record)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Format the log record with ANSI color codes and elapsed time.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>record</code></b>
                  (<code><span title="logging.LogRecord">LogRecord</span></code>)
              –
              <div class="doc-md-description">
                <p>The log record to format.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    str: The formatted log message with ANSI color codes.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.config.logconf.ColoredFormatter.remove_ansi" class="doc doc-heading">
            <code class="highlight language-python">remove_ansi(s)</code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

        <p>Remove ANSI escape codes from a string.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>s</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The string from which to remove ANSI escape codes.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    str: The string without ANSI escape codes.</p>


    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.config.logconf.setup_logger" class="doc doc-heading">
            <code class="highlight language-python">setup_logger(name='phoskintime', log_file=None, level=logging.DEBUG, log_dir=LOG_DIR, rotate=True, max_bytes=2 * 1024 * 1024, backup_count=5)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to set up a logger with both file and console handlers.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>name</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;phoskintime&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Name of the logger.</p>
              </div>
            </li>
            <li>
              <b><code>log_file</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Path to the log file. If None, a default path is generated.</p>
              </div>
            </li>
            <li>
              <b><code>level</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code><span title="logging.DEBUG">DEBUG</span></code>
)
              –
              <div class="doc-md-description">
                <p>Logging level (e.g., logging.DEBUG, logging.INFO).</p>
              </div>
            </li>
            <li>
              <b><code>log_dir</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="kinopt.evol.config.constants.LOG_DIR">LOG_DIR</span></code>
)
              –
              <div class="doc-md-description">
                <p>Directory where log files are stored.</p>
              </div>
            </li>
            <li>
              <b><code>rotate</code></b>
                  (<code><span title="bool">bool</span></code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to use rotating file handler.</p>
              </div>
            </li>
            <li>
              <b><code>max_bytes</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>2 * 1024 * 1024</code>
)
              –
              <div class="doc-md-description">
                <p>Maximum size of log file before rotation.</p>
              </div>
            </li>
            <li>
              <b><code>backup_count</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>5</code>
)
              –
              <div class="doc-md-description">
                <p>Number of backup files to keep.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>logger</code></b>(                  <code><span title="logging.Logger">Logger</span></code>
)              –
              <div class="doc-md-description">
                <p>Configured logger instance.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.exporter.plotout" class="doc doc-heading">
            <code>kinopt.evol.exporter.plotout</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.plotout.plot_residuals_for_gene" class="doc doc-heading">
            <code class="highlight language-python">plot_residuals_for_gene(gene, gene_data)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generates and saves combined residual-related plots for one gene with all psites in the legend.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene identifier.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary with keys 'psites', 'observed', 'estimated', and 'residuals' containing data for all psites.</p>
              </div>
            </li>
            <li>
              <b><code>TIME_POINTS</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span> or <span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points corresponding to the series.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.plotout.opt_analyze_nsga" class="doc doc-heading">
            <code class="highlight language-python">opt_analyze_nsga(problem, result, F, pairs, approx_ideal, approx_nadir, asf_i, pseudo_i, n_evals, hv, hist, val, hist_cv_avg, k, igd, best_objectives, waterfall_df, convergence_df, alpha_values, beta_values)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to generate and save various plots related to optimization results.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>problem</code></b>
              –
              <div class="doc-md-description">
                <p>The optimization problem instance.</p>
              </div>
            </li>
            <li>
              <b><code>result</code></b>
              –
              <div class="doc-md-description">
                <p>The result of the optimization run.</p>
              </div>
            </li>
            <li>
              <b><code>F</code></b>
              –
              <div class="doc-md-description">
                <p>Objective function values.</p>
              </div>
            </li>
            <li>
              <b><code>pairs</code></b>
              –
              <div class="doc-md-description">
                <p>Pairs of objectives to plot.</p>
              </div>
            </li>
            <li>
              <b><code>approx_ideal</code></b>
              –
              <div class="doc-md-description">
                <p>Approximate ideal point in objective space.</p>
              </div>
            </li>
            <li>
              <b><code>approx_nadir</code></b>
              –
              <div class="doc-md-description">
                <p>Approximate nadir point in objective space.</p>
              </div>
            </li>
            <li>
              <b><code>asf_i</code></b>
              –
              <div class="doc-md-description">
                <p>Index of the best solution in terms of the augmented weighted sum.</p>
              </div>
            </li>
            <li>
              <b><code>pseudo_i</code></b>
              –
              <div class="doc-md-description">
                <p>Index of the pseudo weights.</p>
              </div>
            </li>
            <li>
              <b><code>n_evals</code></b>
              –
              <div class="doc-md-description">
                <p>Number of evaluations at each generation.</p>
              </div>
            </li>
            <li>
              <b><code>hv</code></b>
              –
              <div class="doc-md-description">
                <p>Hypervolume values.</p>
              </div>
            </li>
            <li>
              <b><code>hist</code></b>
              –
              <div class="doc-md-description">
                <p>History of the optimization process.</p>
              </div>
            </li>
            <li>
              <b><code>val</code></b>
              –
              <div class="doc-md-description">
                <p>Values for convergence plot.</p>
              </div>
            </li>
            <li>
              <b><code>hist_cv_avg</code></b>
              –
              <div class="doc-md-description">
                <p>Average constraint violation history.</p>
              </div>
            </li>
            <li>
              <b><code>k</code></b>
              –
              <div class="doc-md-description">
                <p>Number of generations.</p>
              </div>
            </li>
            <li>
              <b><code>igd</code></b>
              –
              <div class="doc-md-description">
                <p>Inverted generational distance values.</p>
              </div>
            </li>
            <li>
              <b><code>best_objectives</code></b>
              –
              <div class="doc-md-description">
                <p>Best objectives found during the optimization process.</p>
              </div>
            </li>
            <li>
              <b><code>waterfall_df</code></b>
              –
              <div class="doc-md-description">
                <p>DataFrame containing waterfall plot data.</p>
              </div>
            </li>
            <li>
              <b><code>convergence_df</code></b>
              –
              <div class="doc-md-description">
                <p>DataFrame containing convergence data.</p>
              </div>
            </li>
            <li>
              <b><code>alpha_values</code></b>
              –
              <div class="doc-md-description">
                <p>Dictionary containing alpha values for parameters.</p>
              </div>
            </li>
            <li>
              <b><code>beta_values</code></b>
              –
              <div class="doc-md-description">
                <p>Dictionary containing beta values for parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.plotout.opt_analyze_de" class="doc doc-heading">
            <code class="highlight language-python">opt_analyze_de(long_df, convergence_df, ordered_optimizer_runs, x_values, y_values, val)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to generate and save various plots related to optimization results.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>long_df</code></b>
                  (<code><span title="DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing parameter values and objective function values.</p>
              </div>
            </li>
            <li>
              <b><code>convergence_df</code></b>
                  (<code><span title="DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing convergence data.</p>
              </div>
            </li>
            <li>
              <b><code>ordered_optimizer_runs</code></b>
                  (<code><span title="DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing ordered optimizer runs.</p>
              </div>
            </li>
            <li>
              <b><code>x_values</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>X-axis values for the waterfall plot.</p>
              </div>
            </li>
            <li>
              <b><code>y_values</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Y-axis values for the waterfall plot.</p>
              </div>
            </li>
            <li>
              <b><code>val</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Values for the convergence plot.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.exporter.sheetutils" class="doc doc-heading">
            <code>kinopt.evol.exporter.sheetutils</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.exporter.sheetutils.output_results" class="doc doc-heading">
            <code class="highlight language-python">output_results(P_initial, P_init_dense, P_estimated, residuals, alpha_values, beta_values, result, timepoints, OUT_FILE)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to output results to an Excel file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_initial</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary with initial parameters.</p>
              </div>
            </li>
            <li>
              <b><code>P_init_dense</code></b>
                  (<code><span title="ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Dense matrix of initial parameters.</p>
              </div>
            </li>
            <li>
              <b><code>P_estimated</code></b>
                  (<code><span title="ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Dense matrix of estimated parameters.</p>
              </div>
            </li>
            <li>
              <b><code>residuals</code></b>
                  (<code><span title="ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Dense matrix of residuals.</p>
              </div>
            </li>
            <li>
              <b><code>alpha_values</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary with alpha values.</p>
              </div>
            </li>
            <li>
              <b><code>beta_values</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary with beta values.</p>
              </div>
            </li>
            <li>
              <b><code>result</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Result string for logging.</p>
              </div>
            </li>
            <li>
              <b><code>timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of time points.</p>
              </div>
            </li>
            <li>
              <b><code>OUT_FILE</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Output file path.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.objfn.minfndiffevo" class="doc doc-heading">
            <code>kinopt.evol.objfn.minfndiffevo</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem" class="doc doc-heading">
            <code>PhosphorylationOptimizationProblem</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.ElementwiseProblem">ElementwiseProblem</span></code></p>


        <p>Custom optimization problem for phosphorylation analysis.</p>
<p>Defines the constraints, bounds, and objective function for optimizing
alpha and beta parameters across gene-psite-kinase relationships.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Attributes:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_initial</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </li>
            <li>
              <b><code>P_initial_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </li>
            <li>
              <b><code>K_index</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </li>
            <li>
              <b><code>K_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </li>
            <li>
              <b><code>gene_psite_counts</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </li>
            <li>
              <b><code>beta_counts</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python">__init__(P_initial, P_initial_array, K_index, K_array, gene_psite_counts, beta_counts, **kwargs)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Initializes the optimization problem with given data and constraints.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_initial</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </li>
            <li>
              <b><code>P_initial_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </li>
            <li>
              <b><code>K_index</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </li>
            <li>
              <b><code>K_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </li>
            <li>
              <b><code>gene_psite_counts</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </li>
            <li>
              <b><code>beta_counts</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfndiffevo.PhosphorylationOptimizationProblem.objective_function" class="doc doc-heading">
            <code class="highlight language-python">objective_function(params)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Computes the loss value for the given parameters using the selected loss type.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>params</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Decision variables vector.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>float</code></b>              –
              <div class="doc-md-description">
                <p>Computed loss value.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.objfn.minfnnsgaii" class="doc doc-heading">
            <code>kinopt.evol.objfn.minfnnsgaii</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem" class="doc doc-heading">
            <code>PhosphorylationOptimizationProblem</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.ElementwiseProblem">ElementwiseProblem</span></code></p>


        <p>Multi-objective optimization problem for phosphorylation analysis.</p>
<p>Objectives:
- Minimize sum of squared residuals (main objective).
- Minimize violations of constraints for alpha (secondary objective).
- Minimize violations of constraints for beta (tertiary objective).</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python">__init__(P_initial, P_initial_array, K_index, K_array, gene_psite_counts, beta_counts, **kwargs)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Initializes the multi-objective optimization problem.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_initial</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </li>
            <li>
              <b><code>P_initial_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Observed time-series data for gene-psite pairs.</p>
              </div>
            </li>
            <li>
              <b><code>K_index</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </li>
            <li>
              <b><code>K_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </li>
            <li>
              <b><code>gene_psite_counts</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of kinases per gene-psite combination.</p>
              </div>
            </li>
            <li>
              <b><code>beta_counts</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="kinopt.evol.objfn.minfnnsgaii.PhosphorylationOptimizationProblem.objective_function" class="doc doc-heading">
            <code class="highlight language-python">objective_function(params)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Computes the loss value for the given parameters using the selected loss type.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>params</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Decision variables vector.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>float</code></b>              –
              <div class="doc-md-description">
                <p>Computed loss value.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.opt.optrun" class="doc doc-heading">
            <code>kinopt.evol.opt.optrun</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python">run_optimization(P_initial, P_initial_array, K_index, K_array, gene_psite_counts, beta_counts, PhosphorylationOptimizationProblem)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Sets up and runs the multi-objective optimization problem for phosphorylation
using an NSGA2 algorithm and a thread pool for parallelization.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_initial,</code></b>
                  (<code>(<span title="P_initial_array">P_initial_array</span>, <span title="K_index">K_index</span>, <span title="K_array">K_array</span>, <span title="gene_psite_counts">gene_psite_counts</span>, <span title="beta_counts">beta_counts</span>)</code>)
              –
              <div class="doc-md-description">
                <p>Data structures describing the problem (time-series data, kinases, etc.).</p>
              </div>
            </li>
            <li>
              <b><code>PhosphorylationOptimizationProblem</code></b>
                  (<code>class</code>)
              –
              <div class="doc-md-description">
                <p>The custom problem class to be instantiated.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>result</code></b>              –
              <div class="doc-md-description">
                <p>The pymoo result object containing the optimized population and history.</p>
              </div>
            </li>
            <li>
<b><code>exec_time</code></b>              –
              <div class="doc-md-description">
                <p>Execution time for the optimization.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.opt.optrun.post_optimization_nsga" class="doc doc-heading">
            <code class="highlight language-python">post_optimization_nsga(result, weights=np.array([1.0, 1.0, 1.0]), ref_point=np.array([3, 1, 1]))</code>

</h3>


    <div class="doc doc-contents ">

        <p>Post-processes the result of a multi-objective optimization run.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>result</code></b>
              –
              <div class="doc-md-description">
                <p>The final result object from the optimizer (e.g., a pymoo result).</p>
              </div>
            </li>
            <li>
              <b><code>weights</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>, default:
                      <code><span title="numpy.array">array</span>([1.0, 1.0, 1.0])</code>
)
              –
              <div class="doc-md-description">
                <p>Array of length 3 for weighting the objectives.</p>
              </div>
            </li>
            <li>
              <b><code>ref_point</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>, default:
                      <code><span title="numpy.array">array</span>([3, 1, 1])</code>
)
              –
              <div class="doc-md-description">
                <p>Reference point for hypervolume computations.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary with keys:
'best_solution': The best individual from the weighted scoring.
'best_objectives': Its corresponding objective vector.
'optimized_params': The individual's decision variables (X).
'scores': Weighted scores for each solution in the Pareto front.
'best_index': The index of the best solution according to weighted score.
'hist_hv': The hypervolume per generation.
'hist_igd': The IGD+ per generation.
'convergence_df': The DataFrame with iteration vs. best objective value
    for each iteration in the result history.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.opt.optrun.post_optimization_de" class="doc doc-heading">
            <code class="highlight language-python">post_optimization_de(result, alpha_values, beta_values)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Post-processes the result of a multi-objective optimization run.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>result</code></b>
              –
              <div class="doc-md-description">
                <p>The final result object from the optimizer (e.g., a pymoo result).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary with keys:
'best_solution': The best individual from the weighted scoring.
'best_objectives': Its corresponding objective vector.
'optimized_params': The individual's decision variables (X).
'scores': Weighted scores for each solution in the Pareto front.
'best_index': The index of the best solution according to weighted score.
'hist_hv': The hyper volume per generation.
'hist_igd': The IGD+ per generation.
'convergence_df': The DataFrame with iteration vs. best objective value
    for each iteration in the result history.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.optcon.construct" class="doc doc-heading">
            <code>kinopt.evol.optcon.construct</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.optcon.construct.pipeline" class="doc doc-heading">
            <code class="highlight language-python">pipeline(input1_path: str, input2_path: str, time_series_columns: list[str], scaling_method: str, split_point: float, segment_points: list[float], estimate_missing_kinases: bool, kinase_to_psites: dict[str, int])</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to run the entire pipeline for loading and processing data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>input1_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the first CSV file (HGNC data).</p>
              </div>
            </li>
            <li>
              <b><code>input2_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the second CSV file (kinase interactions).</p>
              </div>
            </li>
            <li>
              <b><code>time_series_columns</code></b>
                  (<code><span title="list">list</span>[<span title="str">str</span>]</code>)
              –
              <div class="doc-md-description">
                <p>List of time series columns to extract.</p>
              </div>
            </li>
            <li>
              <b><code>scaling_method</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Method for scaling the data.</p>
              </div>
            </li>
            <li>
              <b><code>split_point</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Split point for scaling.</p>
              </div>
            </li>
            <li>
              <b><code>segment_points</code></b>
                  (<code><span title="list">list</span>[<span title="float">float</span>]</code>)
              –
              <div class="doc-md-description">
                <p>Segment points for scaling.</p>
              </div>
            </li>
            <li>
              <b><code>estimate_missing_kinases</code></b>
                  (<code><span title="bool">bool</span></code>)
              –
              <div class="doc-md-description">
                <p>Flag to estimate missing kinases.</p>
              </div>
            </li>
            <li>
              <b><code>kinase_to_psites</code></b>
                  (<code><span title="dict">dict</span>[<span title="str">str</span>, <span title="int">int</span>]</code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping kinases to their respective psites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>full_hgnc_df</code></b>(                  <code><span title="pandas.DataFrame">DataFrame</span></code>
)              –
              <div class="doc-md-description">
                <p>The scaled data from input1.</p>
              </div>
            </li>
            <li>
<b><code>interaction_df</code></b>(                  <code><span title="pandas.DataFrame">DataFrame</span></code>
)              –
              <div class="doc-md-description">
                <p>The subset/merged DataFrame from input2.</p>
              </div>
            </li>
            <li>
<b><code>observed</code></b>(                  <code><span title="pandas.DataFrame">DataFrame</span></code>
)              –
              <div class="doc-md-description">
                <p>Subset of full_hgnc_df merged with interaction_df.</p>
              </div>
            </li>
            <li>
<b><code>P_initial</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Dictionary mapping gene-psite pairs to kinase relationships and time-series data.</p>
              </div>
            </li>
            <li>
<b><code>P_initial_array</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Array containing observed time-series data for gene-psite pairs.</p>
              </div>
            </li>
            <li>
<b><code>K_array</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Array containing time-series data for kinase-psite combinations.</p>
              </div>
            </li>
            <li>
<b><code>K_index</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Mapping of kinases to their respective psite data.</p>
              </div>
            </li>
            <li>
<b><code>beta_counts</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Mapping of kinase indices to the number of associated psites.</p>
              </div>
            </li>
            <li>
<b><code>gene_psite_counts</code></b>(                  <code><span title="list">list</span></code>
)              –
              <div class="doc-md-description">
                <p>List of counts of psites for each gene.</p>
              </div>
            </li>
            <li>
<b><code>n</code></b>(                  <code><span title="int">int</span></code>
)              –
              <div class="doc-md-description">
                <p>Number of unique gene-psite pairs.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.optcon.construct.load_geneid_to_psites" class="doc doc-heading">
            <code class="highlight language-python">load_geneid_to_psites(input1_path=INPUT1)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to load geneid to psite mapping from input1.csv.
Args:
    input1_path (str): Path to the first CSV file (HGNC data).
Returns:
    geneid_psite_map (dict): Dictionary mapping gene IDs to sets of psites.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.optcon.construct.get_unique_kinases" class="doc doc-heading">
            <code class="highlight language-python">get_unique_kinases(input2_path=INPUT2)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to extract unique kinases from input2.csv.
Args:
    input2_path (str): Path to the second CSV file (kinase interactions).
Returns:
    kinases (set): Set of unique kinases extracted from the input2 file.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.optcon.construct.check_kinases" class="doc doc-heading">
            <code class="highlight language-python">check_kinases()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to check if kinases from input2.csv are present in input1.csv.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.utils.iodata" class="doc doc-heading">
            <code>kinopt.evol.utils.iodata</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python">format_duration(seconds)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Returns a formatted string representing the duration in seconds, minutes, or hours.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>seconds</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>The duration in seconds.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    str: The formatted duration string.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.load_and_scale_data" class="doc doc-heading">
            <code class="highlight language-python">load_and_scale_data(estimate_missing, scaling_method, split_point, seg_points)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to load and scale data from CSV files.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>estimate_missing</code></b>
                  (<code><span title="bool">bool</span></code>)
              –
              <div class="doc-md-description">
                <p>If True, estimates missing values.</p>
              </div>
            </li>
            <li>
              <b><code>scaling_method</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The scaling method to apply ('min_max', 'log', 'temporal', 'segmented', 'slope', 'cumulative').</p>
              </div>
            </li>
            <li>
              <b><code>split_point</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Column index for temporal scaling.</p>
              </div>
            </li>
            <li>
              <b><code>seg_points</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of column indices for segmented scaling.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>full_hgnc_df</code></b>(                  <code><span title="pandas.DataFrame">DataFrame</span></code>
)              –
              <div class="doc-md-description">
                <p>DataFrame with scaled time-series data.</p>
              </div>
            </li>
            <li>
<b><code>interaction_df</code></b>(                  <code><span title="pandas.DataFrame">DataFrame</span></code>
)              –
              <div class="doc-md-description">
                <p>DataFrame containing interaction data.</p>
              </div>
            </li>
            <li>
<b><code>observed</code></b>(                  <code><span title="pandas.DataFrame">DataFrame</span></code>
)              –
              <div class="doc-md-description">
                <p>DataFrame containing observed data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.apply_scaling" class="doc doc-heading">
            <code class="highlight language-python">apply_scaling(df, time_series_columns, method, split_point, segment_points)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to apply different scaling methods to time-series data in a DataFrame.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>df</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>Input DataFrame containing time-series data.</p>
              </div>
            </li>
            <li>
              <b><code>time_series_columns</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of column names to scale.</p>
              </div>
            </li>
            <li>
              <b><code>method</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Scaling method ('min_max', 'log', 'temporal', 'segmented', 'slope', 'cumulative').</p>
              </div>
            </li>
            <li>
              <b><code>split_point</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Column index for temporal scaling.</p>
              </div>
            </li>
            <li>
              <b><code>segment_points</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of column indices for segmented scaling.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: DataFrame with scaled time-series data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python">create_report(results_dir: str, output_file: str = 'report.html')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>results_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the root result's directory.</p>
              </div>
            </li>
            <li>
              <b><code>output_file</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;report.html&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python">organize_output_files(*directories)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to organize output files into protein-specific folders and a general folder.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>*directories</code></b>
              –
              <div class="doc-md-description">
                <p>List of directories to organize.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.evol.utils.params" class="doc doc-heading">
            <code>kinopt.evol.utils.params</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.params.extract_parameters" class="doc doc-heading">
            <code class="highlight language-python">extract_parameters(P_initial, gene_psite_counts, K_index, optimized_params)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to extract alpha and beta values from the optimized parameters.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_initial</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing initial parameters for each gene-psite pair.</p>
              </div>
            </li>
            <li>
              <b><code>gene_psite_counts</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of counts for each gene-psite pair.</p>
              </div>
            </li>
            <li>
              <b><code>K_index</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping kinases to their respective psite pairs.</p>
              </div>
            </li>
            <li>
              <b><code>optimized_params</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of optimized parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>alpha_values</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Dictionary containing alpha values for each gene-psite pair.</p>
              </div>
            </li>
            <li>
<b><code>beta_values</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Dictionary containing beta values for each kinase-psite pair.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.evol.utils.params.compute_metrics" class="doc doc-heading">
            <code class="highlight language-python">compute_metrics(optimized_params: np.ndarray, P_initial: dict, P_initial_array: np.ndarray, K_index: dict, K_array: np.ndarray, gene_psite_counts: list, beta_counts: dict, n: int)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to compute error metrics for the estimated series.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>optimized_params</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of optimized parameters.</p>
              </div>
            </li>
            <li>
              <b><code>P_initial</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing initial parameters for each gene-psite pair.</p>
              </div>
            </li>
            <li>
              <b><code>P_initial_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of initial parameters.</p>
              </div>
            </li>
            <li>
              <b><code>K_index</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping kinases to their respective psite pairs.</p>
              </div>
            </li>
            <li>
              <b><code>K_array</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of kinases.</p>
              </div>
            </li>
            <li>
              <b><code>gene_psite_counts</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of counts for each gene-psite pair.</p>
              </div>
            </li>
            <li>
              <b><code>beta_counts</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>List of counts for each kinase-psite pair.</p>
              </div>
            </li>
            <li>
              <b><code>n</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of samples.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>P_estimated</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Estimated series.</p>
              </div>
            </li>
            <li>
<b><code>residuals</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Residuals between initial and estimated series.</p>
              </div>
            </li>
            <li>
<b><code>mse</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>Mean Squared Error.</p>
              </div>
            </li>
            <li>
<b><code>rmse</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>Root Mean Squared Error.</p>
              </div>
            </li>
            <li>
<b><code>mae</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>Mean Absolute Error.</p>
              </div>
            </li>
            <li>
<b><code>mape</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>Mean Absolute Percentage Error.</p>
              </div>
            </li>
            <li>
<b><code>r_squared</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>R-squared value.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="gradient-based-algorithms">Gradient-Based Algorithms</h3>


<div class="doc doc-object doc-module">



<h2 id="kinopt.local.config.constants" class="doc doc-heading">
            <code>kinopt.local.config.constants</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python">parse_args()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Parses command-line arguments for the optimization script.
This function uses argparse to handle various parameters related to the optimization process.
The parameters include bounds for the optimization, loss function types, estimation of missing kinases,
scaling methods for time-series data, and the optimization method to be used.
The function returns a tuple containing the parsed arguments.</p>
<p>:return: A tuple containing the parsed arguments.
    - lower_bound (float): Lower bound for the optimization.
    - upper_bound (float): Upper bound for the optimization.
    - loss_type (str): Type of loss function to use.
    - estimate_missing (bool): Whether to estimate missing kinase-psite values.
    - scaling_method (str): Method for scaling time-series data.
    - split_point (int): Split point for temporal scaling.
    - segment_points (list of int): Segment points for segmented scaling.
    - method (str): Optimization method to use.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.config.logconf" class="doc doc-heading">
            <code>kinopt.local.config.logconf</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.exporter.plotout" class="doc doc-heading">
            <code>kinopt.local.exporter.plotout</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_fits_for_gene" class="doc doc-heading">
            <code class="highlight language-python">plot_fits_for_gene(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot the observed and estimated phosphorylation levels for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing observed and estimated data for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_cumulative_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_cumulative_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot the cumulative residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_autocorrelation_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_autocorrelation_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot the autocorrelation of residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_histogram_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_histogram_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot histograms of residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.plotout.plot_qqplot_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_qqplot_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot QQ plots of residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.exporter.sheetutils" class="doc doc-heading">
            <code>kinopt.local.exporter.sheetutils</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_fits_for_gene" class="doc doc-heading">
            <code class="highlight language-python">plot_fits_for_gene(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot the observed and estimated phosphorylation levels for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing observed and estimated data for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_cumulative_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_cumulative_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot the cumulative residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_autocorrelation_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_autocorrelation_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot the autocorrelation of residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_histogram_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_histogram_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot histograms of residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.plot_qqplot_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_qqplot_residuals(gene, gene_data, real_timepoints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot QQ plots of residuals for each psite of a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The name of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>gene_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>A dictionary containing the residuals for each psite of the gene.</p>
              </div>
            </li>
            <li>
              <b><code>real_timepoints</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>A list of timepoints corresponding to the observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.exporter.sheetutils.output_results" class="doc doc-heading">
            <code class="highlight language-python">output_results(P_initial, P_init_dense, P_estimated, residuals, alpha_values, beta_values, result, mse, rmse, mae, mape, r_squared)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to output the results of the optimization process.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_initial</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing initial phosphorylation data.</p>
              </div>
            </li>
            <li>
              <b><code>P_init_dense</code></b>
                  (<code><span title="ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Dense matrix of initial phosphorylation data.</p>
              </div>
            </li>
            <li>
              <b><code>P_estimated</code></b>
                  (<code><span title="ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Dense matrix of estimated phosphorylation data.</p>
              </div>
            </li>
            <li>
              <b><code>residuals</code></b>
                  (<code><span title="ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Dense matrix of residuals.</p>
              </div>
            </li>
            <li>
              <b><code>alpha_values</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing optimized alpha values.</p>
              </div>
            </li>
            <li>
              <b><code>beta_values</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing optimized beta values.</p>
              </div>
            </li>
            <li>
              <b><code>result</code></b>
                  (<code><span title="OptimizeResult">OptimizeResult</span></code>)
              –
              <div class="doc-md-description">
                <p>Result object from the optimization process.</p>
              </div>
            </li>
            <li>
              <b><code>mse</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Mean Squared Error of the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>rmse</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Root Mean Squared Error of the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>mae</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Mean Absolute Error of the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>mape</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Mean Absolute Percentage Error of the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>r_squared</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>R-squared value of the optimization.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.objfn.minfn" class="doc doc-heading">
            <code>kinopt.local.objfn.minfn</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.opt.optrun" class="doc doc-heading">
            <code>kinopt.local.opt.optrun</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python">run_optimization(obj_fun, params_initial, opt_method, bounds, constraints)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Run optimization using the specified method.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>obj_fun</code></b>
              –
              <div class="doc-md-description">
                <p>Objective function to minimize.</p>
              </div>
            </li>
            <li>
              <b><code>params_initial</code></b>
              –
              <div class="doc-md-description">
                <p>Initial parameters for the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>opt_method</code></b>
              –
              <div class="doc-md-description">
                <p>Optimization method to use (e.g., 'SLSQP', 'trust-constr').</p>
              </div>
            </li>
            <li>
              <b><code>bounds</code></b>
              –
              <div class="doc-md-description">
                <p>Bounds for the parameters.</p>
              </div>
            </li>
            <li>
              <b><code>constraints</code></b>
              –
              <div class="doc-md-description">
                <p>Constraints for the optimization.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>result</code></b>              –
              <div class="doc-md-description">
                <p>Result of the optimization.</p>
              </div>
            </li>
            <li>
<b><code>optimized_params</code></b>              –
              <div class="doc-md-description">
                <p>Optimized parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.optcon.construct" class="doc doc-heading">
            <code>kinopt.local.optcon.construct</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.optcon.construct.load_geneid_to_psites" class="doc doc-heading">
            <code class="highlight language-python">load_geneid_to_psites(input1_path=INPUT1)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load the geneid to psite mapping from a CSV file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>input1_path</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="kinopt.local.config.constants.INPUT1">INPUT1</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the input CSV file containing geneid and psite information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    defaultdict: A dictionary mapping geneid to a set of psites.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.optcon.construct.get_unique_kinases" class="doc doc-heading">
            <code class="highlight language-python">get_unique_kinases(input2_path=INPUT2)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Extract unique kinases from the input CSV file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>input2_path</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="kinopt.local.config.constants.INPUT2">INPUT2</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the input CSV file containing kinase information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    set: A set of unique kinases.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.optcon.construct.check_kinases" class="doc doc-heading">
            <code class="highlight language-python">check_kinases()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Check if kinases in input2.csv are present in input1.csv and log the results.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.utils.iodata" class="doc doc-heading">
            <code>kinopt.local.utils.iodata</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python">format_duration(seconds)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Formats a duration in seconds into a human-readable string.
- If less than 60 seconds, returns in seconds.
- If less than 3600 seconds, returns in minutes.
- If more than 3600 seconds, returns in hours.</p>
<p>:param seconds:
:return: Formatted string</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.load_and_scale_data" class="doc doc-heading">
            <code class="highlight language-python">load_and_scale_data(estimate_missing, scaling_method, split_point, seg_points)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load and scale the data from the specified input files.</p>
<p>:param estimate_missing:
:param scaling_method:
:param split_point:
:param seg_points:
:return: Time series data, interaction data, observed data</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.apply_scaling" class="doc doc-heading">
            <code class="highlight language-python">apply_scaling(df, cols, method, split_point, seg_points)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Apply scaling to the specified columns of a DataFrame based on the given method.
The scaling methods include:
- 'min_max': Min-Max scaling
- 'log': Logarithmic scaling
- 'temporal': Temporal scaling (two segments)
- 'segmented': Segmented scaling (multiple segments)
- 'slope': Slope scaling
- 'cumulative': Cumulative scaling</p>
<p>:param df:
:param cols:
:param method:
:param split_point:
:param seg_points:
:return: df</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python">create_report(results_dir: str, output_file: str = 'report.html')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>
<p>For each gene folder (e.g. "ABL2"), the report will include:
  - All PNG plots and interactive HTML plots displayed in a grid with three plots per row.
  - Each plot is confined to a fixed size of 900px by 900px.
  - Data tables from XLSX or CSV files in the gene folder are displayed below the plots, one per row.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>results_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </li>
            <li>
              <b><code>output_file</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;report.html&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python">organize_output_files(*directories)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to organize output files into protein-specific folders.
It moves files matching the pattern 'protein_name_*.{json,svg,png,html,csv,xlsx}'
into a folder named after the protein (e.g., 'ABL2') and moves all other files
into a 'General' folder within the same directory.</p>
<p>:param directories:</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.local.utils.params" class="doc doc-heading">
            <code>kinopt.local.utils.params</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.params.extract_parameters" class="doc doc-heading">
            <code class="highlight language-python">extract_parameters(P_initial, gene_kinase_counts, total_alpha, unique_kinases, K_index, optimized_params)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Extracts the alpha and beta parameters from the optimized parameters.</p>
<p>:param P_initial:
:param gene_kinase_counts:
:param total_alpha:
:param unique_kinases:
:param K_index:
:param optimized_params:
:return: Alpha and beta values as dictionaries</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.local.utils.params.compute_metrics" class="doc doc-heading">
            <code class="highlight language-python">compute_metrics(optimized_params, P_init_dense, t_max, gene_alpha_starts, gene_kinase_counts, gene_kinase_idx, total_alpha, kinase_beta_starts, kinase_beta_counts, K_data, K_indices, K_indptr)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Computes the estimated series and various metrics based on the optimized parameters.</p>
<p>:param optimized_params:
:param P_init_dense:
:param t_max:
:param gene_alpha_starts:
:param gene_kinase_counts:
:param gene_kinase_idx:
:param total_alpha:
:param kinase_beta_starts:
:param kinase_beta_counts:
:param K_data:
:param K_indices:
:param K_indptr:
:return: Estimated series, residuals, MSE, RMSE, MAE, MAPE, R-squared</p>


    </div>

</div>



  </div>

    </div>

</div><h2 id="fitting-analysis-feasibility">Fitting Analysis &amp; Feasibility</h2>


<div class="doc doc-object doc-module">



<h2 id="kinopt.fitanalysis.helpers.postfit" class="doc doc-heading">
            <code>kinopt.fitanalysis.helpers.postfit</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.goodnessoffit" class="doc doc-heading">
            <code class="highlight language-python">goodnessoffit(estimated, observed)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot the goodness of fit and kullback-leibler divergence
for estimated and observed values.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>estimated</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing estimated values.</p>
              </div>
            </li>
            <li>
              <b><code>observed</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing observed values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.reshape_alpha_beta" class="doc doc-heading">
            <code class="highlight language-python">reshape_alpha_beta(alpha_values, beta_values)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to reshape alpha and beta values for plotting.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>alpha_values</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing alpha values.</p>
              </div>
            </li>
            <li>
              <b><code>beta_values</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing beta values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    pd.DataFrame: Reshaped DataFrame containing both alpha and beta values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.perform_pca" class="doc doc-heading">
            <code class="highlight language-python">perform_pca(df)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to perform PCA analysis on the given DataFrame.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>df</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing the data for PCA analysis.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: DataFrame with PCA results and additional columns for type and gene/psite information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.plot_pca" class="doc doc-heading">
            <code class="highlight language-python">plot_pca(result_df_sorted, y_axis_column)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Plot PCA or t-SNE results for each gene/psite.
The function creates scatter plots with different markers for alpha and beta parameters,
and adds labels for each point.
The function also adjusts text labels to avoid overlap using the adjustText library.</p>
<p>:param result_df_sorted: DataFrame containing PCA or t-SNE results.
:param y_axis_column: Column name for the y-axis values in the plot.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.perform_tsne" class="doc doc-heading">
            <code class="highlight language-python">perform_tsne(scaled_data, df)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Perform t-SNE analysis on the given scaled data.
The function returns a DataFrame with t-SNE results and additional columns for type and gene/psite information.</p>
<p>:param scaled_data:
:param df:</p>
<p>:return:
- pd.DataFrame: DataFrame with t-SNE results and additional columns.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.additional_plots" class="doc doc-heading">
            <code class="highlight language-python">additional_plots(df, scaled_data, alpha_values, beta_values, residuals_df)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to create additional plots including CDF, KDE, Boxplot, and Hierarchical Clustering.</p>
<p>:param df:
:param scaled_data:
:param alpha_values:
:param beta_values:
:param residuals_df:</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.create_sankey_from_network" class="doc doc-heading">
            <code class="highlight language-python">create_sankey_from_network(output_dir, data, title)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a Sankey diagram from the given data and saves it as an HTML file.</p>
<p>This function processes the input data to generate nodes and links for a Sankey diagram.
It assigns colors to nodes and links based on their attributes and values, and uses Plotly
to render the diagram. The resulting diagram is saved as an HTML file in the specified output directory.</p>
<p>:param output_dir: str
    The directory where the Sankey diagram HTML file will be saved.
:param data: pd.DataFrame
    A DataFrame containing the data for the Sankey diagram. It must include the following columns:
    - 'Source': The source node of the link.
    - 'Target': The target node of the link.
    - 'Value': The value of the link, which determines the flow size.
:param title: str
    The title of the Sankey diagram.</p>
<p>The function performs the following steps:
1. Initializes nodes and links for the Sankey diagram.
2. Maps node labels to indices and assigns colors to nodes.
3. Processes the data to create links between nodes, assigning colors based on link values.
4. Builds the Sankey diagram using Plotly.
5. Adds a color bar to explain the flow gradient.
6. Saves the Sankey diagram as an HTML file in the specified output directory.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.fitanalysis.helpers.postfit.important_connections" class="doc doc-heading">
            <code class="highlight language-python">important_connections(output_dir, data, top_n=20)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Extracts the top N most important connections based on their absolute values
and saves them to a CSV file.</p>
<p>:param output_dir: str
    The directory where the CSV file will be saved.
:param data: pd.DataFrame
    A DataFrame containing the connections with columns 'Source', 'Target', and 'Value'.
:param top_n: int, optional
    The number of top connections to extract (default is 20).</p>
<p>The function sorts the connections by their absolute values in descending order,
selects the top N connections, and saves them to a CSV file named 'top_connections.csv'
in the specified output directory.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="kinopt.optimality.KKT" class="doc doc-heading">
            <code>kinopt.optimality.KKT</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.generate_latex_table" class="doc doc-heading">
            <code class="highlight language-python">generate_latex_table(summary_dict, table_caption, table=None)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to generate a LaTeX table from a summary dictionary.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>summary_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing summary data.</p>
              </div>
            </li>
            <li>
              <b><code>table_caption</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Caption for the LaTeX table.</p>
              </div>
            </li>
            <li>
              <b><code>table</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Optional existing LaTeX table to append to.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>str</code></b>              –
              <div class="doc-md-description">
                <p>LaTeX formatted table as a string.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.print_primal_feasibility_results" class="doc doc-heading">
            <code class="highlight language-python">print_primal_feasibility_results(primal_summary, alpha_violations, beta_violations, logger_obj=None)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Logs the primal feasibility summary and violation details.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>primal_summary</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing primal feasibility results.</p>
              </div>
            </li>
            <li>
              <b><code>alpha_violations</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing alpha constraint violations.</p>
              </div>
            </li>
            <li>
              <b><code>beta_violations</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing beta constraint violations.</p>
              </div>
            </li>
            <li>
              <b><code>logger_obj</code></b>
              –
              <div class="doc-md-description">
                <p>Optional logger object to log the information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.print_sensitivity_and_active_constraints" class="doc doc-heading">
            <code class="highlight language-python">print_sensitivity_and_active_constraints(sensitivity_summary, active_constraints_summary, logger_obj=None)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Logs the sensitivity summary and active constraints summary.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>sensitivity_summary</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing sensitivity analysis results.</p>
              </div>
            </li>
            <li>
              <b><code>active_constraints_summary</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing active constraints summary.</p>
              </div>
            </li>
            <li>
              <b><code>logger_obj</code></b>
              –
              <div class="doc-md-description">
                <p>Optional logger object to log the information.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.plot_constraint_violations" class="doc doc-heading">
            <code class="highlight language-python">plot_constraint_violations(alpha_violations, beta_violations, out_dir)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot constraint violations for alpha and beta values.
It creates a stacked bar plot showing the violations for each protein.
The top 5 proteins with the highest violations are highlighted in red.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>alpha_violations</code></b>
                  (<code><span title="pandas.Series">Series</span></code>)
              –
              <div class="doc-md-description">
                <p>Series containing alpha constraint violations.</p>
              </div>
            </li>
            <li>
              <b><code>beta_violations</code></b>
                  (<code><span title="pandas.Series">Series</span></code>)
              –
              <div class="doc-md-description">
                <p>Series containing beta constraint violations.</p>
              </div>
            </li>
            <li>
              <b><code>out_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Directory to save the plot.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.plot_sensitivity_analysis" class="doc doc-heading">
            <code class="highlight language-python">plot_sensitivity_analysis(sensitivity_analysis, out_dir)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to plot sensitivity analysis results.
It creates a horizontal bar plot showing the mean, max, and min sensitivity for each protein.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>sensitivity_analysis</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing sensitivity analysis results.</p>
              </div>
            </li>
            <li>
              <b><code>out_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Directory to save the plot.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.process_excel_results" class="doc doc-heading">
            <code class="highlight language-python">process_excel_results(file_path=OUT_FILE)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to process the Excel results file.
It reads the alpha and beta values, estimated and observed values,
validates normalization constraints, computes residuals and gradients,
and generates LaTeX tables for the residuals and sensitivity summaries.
It also performs sensitivity analysis and identifies high sensitivity sites.
The results are returned as a dictionary.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>file_path</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="kinopt.evol.config.constants.OUT_FILE">OUT_FILE</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the Excel file containing results.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    dict: Dictionary containing the processed results, including alpha and beta values,
          estimated and observed values, constraint violations, residuals summary,
          sensitivity summary, and high sensitivity sites.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="kinopt.optimality.KKT.post_optimization_results" class="doc doc-heading">
            <code class="highlight language-python">post_optimization_results()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to process and visualize the results of the optimization.</p>
        <p>Returns:
    dict: Dictionary containing the processed results, including alpha and beta values,
          estimated and observed values, constraint violations, residuals summary,
          sensitivity summary, and high sensitivity sites.</p>


    </div>

</div>



  </div>

    </div>

</div><h2 id="tf-mrna-optimization">TF-mRNA Optimization</h2>
<h3 id="evolutionary-algorithms_1">Evolutionary Algorithms</h3>


<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.config.constants" class="doc doc-heading">
            <code>tfopt.evol.config.constants</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python">parse_args()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Parse command line arguments for the PhosKinTime optimization problem.
This function uses argparse to handle input parameters for the optimization process.
The parameters include:
- lower_bound: Lower bound for the optimization variables (default: -2).
- upper_bound: Upper bound for the optimization variables (default: 2).
- loss_type: Type of loss function to use (default: 0).
    Options:
    0: MSE
    1: MAE
    2: soft L1
    3: Cauchy
    4: Arctan
    5: Elastic Net
    6: Tikhonov
- optimizer: Global Evolutionary Optimization method (default: 0).
    Options:
    0: NGSA2
    1: SMSEMOA
    2: AGEMOEA</p>
<p>:returns
- lower_bound: Lower bound for the optimization variables.
- upper_bound: Upper bound for the optimization variables.
- loss_type: Type of loss function to use.
- optimizer: Global Evolutionary Optimization method.
:rtype: tuple
:raises argparse.ArgumentError: If an invalid argument is provided.
:raises SystemExit: If the script is run with invalid arguments.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.config.logconf" class="doc doc-heading">
            <code>tfopt.evol.config.logconf</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.exporter.plotout" class="doc doc-heading">
            <code>tfopt.evol.exporter.plotout</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.exporter.plotout.plot_estimated_vs_observed" class="doc doc-heading">
            <code class="highlight language-python">plot_estimated_vs_observed(predictions, expression_matrix, gene_ids, time_points, regulators, tf_protein_matrix, tf_ids, num_targets, save_path=OUT_DIR)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Plot the estimated vs observed expression levels for a set of genes.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>predictions</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Predicted expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>expression_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Observed expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>gene_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of gene identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points for the experiments.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of regulators for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of TF protein levels.</p>
              </div>
            </li>
            <li>
              <b><code>tf_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>num_targets</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of target genes to plot.</p>
              </div>
            </li>
            <li>
              <b><code>save_path</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.evol.config.constants.OUT_DIR">OUT_DIR</span></code>
)
              –
              <div class="doc-md-description">
                <p>Directory to save the plots.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.exporter.plotout.compute_predictions" class="doc doc-heading">
            <code class="highlight language-python">compute_predictions(x, regulators, protein_mat, psite_tensor, n_reg, T_use, n_mRNA, beta_start_indices, num_psites)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Compute the predicted expression levels based on the optimization variables.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Optimization variables.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of regulators for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>protein_mat</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of TF protein levels.</p>
              </div>
            </li>
            <li>
              <b><code>psite_tensor</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Tensor of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of time points to use.</p>
              </div>
            </li>
            <li>
              <b><code>n_mRNA</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of mRNAs.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of starting indices for beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.exporter.sheetutils" class="doc doc-heading">
            <code>tfopt.evol.exporter.sheetutils</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.exporter.sheetutils.save_results_to_excel" class="doc doc-heading">
            <code class="highlight language-python">save_results_to_excel(gene_ids, tf_ids, final_alpha, final_beta, psite_labels_arr, expression_matrix, predictions, objective_value, reg_map, filename=OUT_FILE)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Save the optimization results to an Excel file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of gene identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>tf_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>final_alpha</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Final alpha values.</p>
              </div>
            </li>
            <li>
              <b><code>final_beta</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Final beta values.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_arr</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of phosphorylation site labels.</p>
              </div>
            </li>
            <li>
              <b><code>expression_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Observed expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>predictions</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Predicted expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>objective_value</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Objective value from optimization.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of genes to regulators.</p>
              </div>
            </li>
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.evol.config.constants.OUT_FILE">OUT_FILE</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the output Excel file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.objfn.minfn" class="doc doc-heading">
            <code>tfopt.evol.objfn.minfn</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem" class="doc doc-heading">
            <code>TFOptimizationMultiObjectiveProblem</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="pymoo.core.problem.Problem">Problem</span></code></p>


        <p>Originally implemented by Julius Normann.</p>
<p>This version has been modified and optimized
for consistency &amp; speed in submodules by Abhinav Mishra.</p>
<p>Multi-objective optimization problem for TF optimization.
This class defines a multi-objective optimization problem for the
transcription factor (TF) optimization problem. It inherits from the
<code>Problem</code> class in the pymoo library. The problem is defined with three
objectives: f1 (error), f2 (alpha violation), and f3 (beta violation).</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="tfopt.evol.objfn.minfn.TFOptimizationMultiObjectiveProblem.__init__" class="doc doc-heading">
            <code class="highlight language-python">__init__(n_var, n_mRNA, n_TF, n_reg, n_psite_max, n_alpha, mRNA_mat, regulators, protein_mat, psite_tensor, T_use, beta_start_indices, num_psites, no_psite_tf, xl=None, xu=None, **kwargs)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Initialize the multi-objective optimization problem.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>n_var</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of decision variables.</p>
              </div>
            </li>
            <li>
              <b><code>n_mRNA</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of mRNAs.</p>
              </div>
            </li>
            <li>
              <b><code>n_TF</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of transcription factors.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>n_psite_max</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Maximum number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>n_alpha</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </li>
            <li>
              <b><code>mRNA_mat</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of mRNA measurements.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of regulators for each mRNA.</p>
              </div>
            </li>
            <li>
              <b><code>protein_mat</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of TF protein levels.</p>
              </div>
            </li>
            <li>
              <b><code>psite_tensor</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Tensor of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of time points to use.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of starting indices for beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>no_psite_tf</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List indicating if a TF has no phosphorylation site.</p>
              </div>
            </li>
            <li>
              <b><code>xl</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Lower bounds for decision variables. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>xu</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Upper bounds for decision variables. Defaults to None.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.objfn.minfn.objective_" class="doc doc-heading">
            <code class="highlight language-python">objective_(x, mRNA_mat, regulators, protein_mat, psite_tensor, n_reg, T_use, n_mRNA, beta_start_indices, num_psites, loss_type, lam1=0.001, lam2=0.001)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Computes a loss value for transcription factor optimization using evolutionary algorithms.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Optimization variables.</p>
              </div>
            </li>
            <li>
              <b><code>mRNA_mat</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of mRNA measurements.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of regulators for each mRNA.</p>
              </div>
            </li>
            <li>
              <b><code>protein_mat</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of TF protein levels.</p>
              </div>
            </li>
            <li>
              <b><code>psite_tensor</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Tensor of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of time points to use.</p>
              </div>
            </li>
            <li>
              <b><code>n_mRNA</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of mRNAs.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of starting indices for beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>loss_type</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Type of loss function to use.</p>
              </div>
            </li>
            <li>
              <b><code>lam1</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>0.001</code>
)
              –
              <div class="doc-md-description">
                <p>L1 penalty coefficient. Defaults to 1e-3.</p>
              </div>
            </li>
            <li>
              <b><code>lam2</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>0.001</code>
)
              –
              <div class="doc-md-description">
                <p>L2 penalty coefficient. Defaults to 1e-3.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>float</code></b>              –
              <div class="doc-md-description">
                <p>Computed loss value.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.opt.optrun" class="doc doc-heading">
            <code>tfopt.evol.opt.optrun</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.opt.optrun.run_optimization" class="doc doc-heading">
            <code class="highlight language-python">run_optimization(problem, total_dim, optimizer)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Run the optimization using the specified algorithm and problem.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>problem</code></b>
                  (<code><span title="Problem">Problem</span></code>)
              –
              <div class="doc-md-description">
                <p>The optimization problem to solve.</p>
              </div>
            </li>
            <li>
              <b><code>total_dim</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Total number of dimensions in the problem.</p>
              </div>
            </li>
            <li>
              <b><code>optimizer</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>The optimizer to use (0 for NSGA2, 1 for SMSEMOA, 2 for AGEMOEA).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>res</code></b>(                  <code><span title="Result">Result</span></code>
)              –
              <div class="doc-md-description">
                <p>The result of the optimization.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.optcon.construct" class="doc doc-heading">
            <code>tfopt.evol.optcon.construct</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.construct.build_fixed_arrays" class="doc doc-heading">
            <code class="highlight language-python">build_fixed_arrays(mRNA_ids, mRNA_mat, TF_ids, protein_dict, psite_dict, psite_labels_dict, reg_map)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Builds fixed-shape arrays from the input data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mRNA_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of mRNA identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>mRNA_mat</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of mRNA expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>TF_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>protein_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TFs to their protein levels.</p>
              </div>
            </li>
            <li>
              <b><code>psite_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TFs to their phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TFs to their phosphorylation site labels.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of genes to their regulators.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    mRNA_mat (np.ndarray): Matrix of mRNA expression levels.
    regulators (np.ndarray): Matrix of regulators for each mRNA.
    protein_mat (np.ndarray): Matrix of TF protein levels.
    psite_tensor (np.ndarray): Tensor of phosphorylation sites.
    n_reg (int): Number of regulators.
    n_psite_max (int): Maximum number of phosphorylation sites across all TFs.
    psite_labels_arr (list): List of phosphorylation site labels for each TF.
    num_psites (np.ndarray): Array indicating the number of phosphorylation sites for each TF.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.optcon.filter" class="doc doc-heading">
            <code>tfopt.evol.optcon.filter</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.load_raw_data" class="doc doc-heading">
            <code class="highlight language-python">load_raw_data()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load raw data from files.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>mRNA_ids</code></b>              –
              <div class="doc-md-description">
                <p>List of mRNA gene identifiers.</p>
              </div>
            </li>
            <li>
<b><code>mRNA_mat</code></b>              –
              <div class="doc-md-description">
                <p>Matrix of mRNA expression data.</p>
              </div>
            </li>
            <li>
<b><code>mRNA_time_cols</code></b>              –
              <div class="doc-md-description">
                <p>Time points for mRNA data.</p>
              </div>
            </li>
            <li>
<b><code>TF_ids</code></b>              –
              <div class="doc-md-description">
                <p>List of transcription factor identifiers.</p>
              </div>
            </li>
            <li>
<b><code>protein_dict</code></b>              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF_ids to their protein data.</p>
              </div>
            </li>
            <li>
<b><code>psite_dict</code></b>              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF_ids to their phosphorylation site data.</p>
              </div>
            </li>
            <li>
<b><code>psite_labels_dict</code></b>              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF_ids to their phosphorylation site labels.</p>
              </div>
            </li>
            <li>
<b><code>TF_time_cols</code></b>              –
              <div class="doc-md-description">
                <p>Time points for TF data.</p>
              </div>
            </li>
            <li>
<b><code>reg_map</code></b>              –
              <div class="doc-md-description">
                <p>Regulation map, mapping mRNA genes to their regulators.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.filter_mrna" class="doc doc-heading">
            <code class="highlight language-python">filter_mrna(mRNA_ids, mRNA_mat, reg_map)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Filter mRNA genes to only those with regulators present in the regulation map.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mRNA_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of mRNA gene identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>mRNA_mat</code></b>
                  (<code><span title="np.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of mRNA expression data.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Regulation map, mapping mRNA genes to their regulators.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>filtered_mRNA_ids</code></b>(                  <code><span title="list">list</span></code>
)              –
              <div class="doc-md-description">
                <p>List of filtered mRNA gene identifiers.</p>
              </div>
            </li>
            <li>
<b><code>filtered_mRNA_mat</code></b>(                  <code><span title="np.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Matrix of filtered mRNA expression data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.update_regulations" class="doc doc-heading">
            <code class="highlight language-python">update_regulations(mRNA_ids, reg_map, TF_ids)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Update the regulation map to only include relevant transcription factors.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mRNA_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of mRNA gene identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Regulation map, mapping mRNA genes to their regulators.</p>
              </div>
            </li>
            <li>
              <b><code>TF_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of transcription factor identifiers.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>relevant_TFs</code></b>(                  <code><span title="set">set</span></code>
)              –
              <div class="doc-md-description">
                <p>Set of relevant transcription factors.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.filter_TF" class="doc doc-heading">
            <code class="highlight language-python">filter_TF(TF_ids, protein_dict, psite_dict, psite_labels_dict, relevant_TFs)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Filter transcription factors to only those present in the relevant_TFs set.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>TF_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of transcription factor identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>protein_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF_ids to their protein data.</p>
              </div>
            </li>
            <li>
              <b><code>psite_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF_ids to their phosphorylation site data.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF_ids to their phosphorylation site labels.</p>
              </div>
            </li>
            <li>
              <b><code>relevant_TFs</code></b>
                  (<code><span title="set">set</span></code>)
              –
              <div class="doc-md-description">
                <p>Set of relevant transcription factors.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>TF_ids_filtered</code></b>(                  <code><span title="list">list</span></code>
)              –
              <div class="doc-md-description">
                <p>List of filtered transcription factor identifiers.</p>
              </div>
            </li>
            <li>
<b><code>protein_dict</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Filtered dictionary mapping TF_ids to their protein data.</p>
              </div>
            </li>
            <li>
<b><code>psite_dict</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Filtered dictionary mapping TF_ids to their phosphorylation site data.</p>
              </div>
            </li>
            <li>
<b><code>psite_labels_dict</code></b>(                  <code><span title="dict">dict</span></code>
)              –
              <div class="doc-md-description">
                <p>Filtered dictionary mapping TF_ids to their phosphorylation site labels.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.optcon.filter.determine_T_use" class="doc doc-heading">
            <code class="highlight language-python">determine_T_use(mRNA_mat, TF_time_cols)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Determine the number of time points to use for the analysis.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mRNA_mat</code></b>
                  (<code><span title="np.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of mRNA expression data.</p>
              </div>
            </li>
            <li>
              <b><code>TF_time_cols</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points for TF data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.utils.iodata" class="doc doc-heading">
            <code>tfopt.evol.utils.iodata</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.load_mRNA_data" class="doc doc-heading">
            <code class="highlight language-python">load_mRNA_data(filename=INPUT3)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load mRNA data from a CSV file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.evol.config.constants.INPUT3">INPUT3</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the CSV file containing mRNA data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    - mRNA_ids: List of mRNA gene identifiers (strings).
    - mRNA_mat: Matrix of mRNA expression data (numpy array).
    - time_cols: List of time columns (excluding "GeneID").</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.load_TF_data" class="doc doc-heading">
            <code class="highlight language-python">load_TF_data(filename=INPUT1)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load TF data from a CSV file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.evol.config.constants.INPUT1">INPUT1</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the CSV file containing TF data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    - TF_ids: List of TF identifiers (strings).
    - protein_dict: Dictionary mapping TF identifiers to their protein data (numpy array).
    - psite_dict: Dictionary mapping TF identifiers to their phosphorylation site data (list of numpy arrays).
    - psite_labels_dict: Dictionary mapping TF identifiers to their phosphorylation site labels (list of strings).
    - time_cols: List of time columns (excluding "GeneID" and "Psite").</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.load_regulation" class="doc doc-heading">
            <code class="highlight language-python">load_regulation(filename=INPUT4)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load regulation data from a CSV file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.evol.config.constants.INPUT4">INPUT4</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the CSV file containing regulation data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    - reg_map: Dictionary mapping mRNA genes to their regulators (list of TF identifiers).</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python">create_report(results_dir: str, output_file: str = 'report.html')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>results_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the directory containing gene folders.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python">organize_output_files(*directories)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Organizes output files from multiple directories into separate folders for each protein.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>directories</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>()</code>
)
              –
              <div class="doc-md-description">
                <p>List of directories to organize.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.iodata.format_duration" class="doc doc-heading">
            <code class="highlight language-python">format_duration(seconds)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Format a duration in seconds into a human-readable string.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>seconds</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Duration in seconds.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    str: Formatted duration string.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.evol.utils.params" class="doc doc-heading">
            <code>tfopt.evol.utils.params</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.create_no_psite_array" class="doc doc-heading">
            <code class="highlight language-python">create_no_psite_array(n_TF, num_psites, psite_labels_arr)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Create an array indicating whether each TF has no phosphorylation sites.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>n_TF</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of transcription factors.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_arr</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of phosphorylation site labels for each TF.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>no_psite_tf</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Array indicating whether each TF has no phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.compute_beta_indices" class="doc doc-heading">
            <code class="highlight language-python">compute_beta_indices(num_psites, n_TF)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Compute the starting indices for the beta parameters for each TF.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>n_TF</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of transcription factors.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>beta_start_indices</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Array of starting indices for the beta parameters.</p>
              </div>
            </li>
            <li>
<b><code>cum</code></b>(                  <code><span title="int">int</span></code>
)              –
              <div class="doc-md-description">
                <p>Total number of beta parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.create_initial_guess" class="doc doc-heading">
            <code class="highlight language-python">create_initial_guess(n_mRNA, n_reg, n_TF, num_psites, no_psite_tf)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Create the initial guess for the optimization variables.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>n_mRNA</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of mRNAs.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>n_TF</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of transcription factors.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>no_psite_tf</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array indicating whether each TF has no phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>x0</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Initial guess for the optimization variables.</p>
              </div>
            </li>
            <li>
<b><code>n_alpha</code></b>(                  <code><span title="int">int</span></code>
)              –
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.create_bounds" class="doc doc-heading">
            <code class="highlight language-python">create_bounds(n_alpha, n_beta_total, lb, ub)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Create the lower and upper bounds for the optimization variables.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>n_alpha</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </li>
            <li>
              <b><code>n_beta_total</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Total number of beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>lb</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Lower bound for the optimization variables.</p>
              </div>
            </li>
            <li>
              <b><code>ub</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Upper bound for the optimization variables.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>xl</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Lower bounds for the optimization variables.</p>
              </div>
            </li>
            <li>
<b><code>xu</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Upper bounds for the optimization variables.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.get_parallel_runner" class="doc doc-heading">
            <code class="highlight language-python">get_parallel_runner()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Get a parallel runner for multi-threading.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>runner</code></b>              –
              <div class="doc-md-description">
                <p>Parallelization runner.</p>
              </div>
            </li>
            <li>
<b><code>pool</code></b>              –
              <div class="doc-md-description">
                <p>ThreadPool instance for parallel execution.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.extract_best_solution" class="doc doc-heading">
            <code class="highlight language-python">extract_best_solution(res, n_alpha, n_mRNA, n_reg, n_TF, num_psites, beta_start_indices)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Extract the best solution from the optimization results.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>res</code></b>
              –
              <div class="doc-md-description">
                <p>Optimization results.</p>
              </div>
            </li>
            <li>
              <b><code>n_alpha</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </li>
            <li>
              <b><code>n_mRNA</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of mRNAs.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>n_TF</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of transcription factors.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of starting indices for the beta parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>final_alpha</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Final alpha parameters.</p>
              </div>
            </li>
            <li>
<b><code>final_beta</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Final beta parameters.</p>
              </div>
            </li>
            <li>
<b><code>best_objectives</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Best objectives from the Pareto front.</p>
              </div>
            </li>
            <li>
<b><code>final_x</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Final optimization variables.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.print_alpha_mapping" class="doc doc-heading">
            <code class="highlight language-python">print_alpha_mapping(mRNA_ids, reg_map, TF_ids, final_alpha)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Print the mapping of transcription factors (TFs) to mRNAs with their corresponding alpha values.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mRNA_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of mRNA identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of genes to their regulators.</p>
              </div>
            </li>
            <li>
              <b><code>TF_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>final_alpha</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Final alpha parameters (mRNA x TF).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.evol.utils.params.print_beta_mapping" class="doc doc-heading">
            <code class="highlight language-python">print_beta_mapping(TF_ids, final_beta, psite_labels_arr)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Print the mapping of transcription factors (TFs) to their beta parameters.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>TF_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>final_beta</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Final beta parameters (TF x β).</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_arr</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of phosphorylation site labels for each TF.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="gradient-based-algorithms_1">Gradient-Based Algorithms</h3>


<div class="doc doc-object doc-module">



<h2 id="tfopt.local.config.constants" class="doc doc-heading">
            <code>tfopt.local.config.constants</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.config.constants.parse_args" class="doc doc-heading">
            <code class="highlight language-python">parse_args()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Parse command line arguments for the PhosKinTime optimization problem.
This function uses argparse to handle input parameters for the optimization process.
The parameters include:
- lower_bound: Lower bound for the optimization variables (default: -2).
- upper_bound: Upper bound for the optimization variables (default: 2).
- loss_type: Type of loss function to use (default: 0).
    Options:
    0: MSE
    1: MAE
    2: soft L1
    3: Cauchy
    4: Arctan
    5: Elastic Net
    6: Tikhonov</p>
<p>:return: lower_bound, upper_bound, loss_type</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.config.logconf" class="doc doc-heading">
            <code>tfopt.local.config.logconf</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.exporter.plotout" class="doc doc-heading">
            <code>tfopt.local.exporter.plotout</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.exporter.plotout.plot_estimated_vs_observed" class="doc doc-heading">
            <code class="highlight language-python">plot_estimated_vs_observed(predictions, expression_matrix, gene_ids, time_points, regulators, tf_protein_matrix, tf_ids, num_targets, save_path=OUT_DIR)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Plots the estimated vs observed values for a given set of genes and their corresponding TFs.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>predictions</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Predicted expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>expression_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Observed expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>gene_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of gene identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points for the experiments.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of regulators for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Matrix of TF protein levels.</p>
              </div>
            </li>
            <li>
              <b><code>tf_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>num_targets</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of target genes to plot.</p>
              </div>
            </li>
            <li>
              <b><code>save_path</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.OUT_DIR">OUT_DIR</span></code>
)
              –
              <div class="doc-md-description">
                <p>Directory to save the plots.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.exporter.sheetutils" class="doc doc-heading">
            <code>tfopt.local.exporter.sheetutils</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.exporter.sheetutils.save_results_to_excel" class="doc doc-heading">
            <code class="highlight language-python">save_results_to_excel(gene_ids, tf_ids, final_alpha, final_beta, psite_labels_arr, expression_matrix, predictions, objective_value, reg_map, filename=OUT_FILE)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Save the optimization results to an Excel file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of gene identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>tf_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>final_alpha</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Final alpha values.</p>
              </div>
            </li>
            <li>
              <b><code>final_beta</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Final beta values.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_arr</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of phosphorylation site labels.</p>
              </div>
            </li>
            <li>
              <b><code>expression_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Observed expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>predictions</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Predicted expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>objective_value</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Objective value from optimization.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Mapping of genes to regulators.</p>
              </div>
            </li>
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.OUT_FILE">OUT_FILE</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the output Excel file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.objfn.minfn" class="doc doc-heading">
            <code>tfopt.local.objfn.minfn</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.objfn.minfn.objective_" class="doc doc-heading">
            <code class="highlight language-python">objective_(x, expression_matrix, regulators, tf_protein_matrix, psite_tensor, n_reg, T_use, n_genes, beta_start_indices, num_psites, loss_type, lam1=1e-06, lam2=1e-06)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Originally implemented by Julius Normann.</p>
<p>This version has been modified and optimized
for consistency &amp; speed in submodules by Abhinav Mishra.</p>
<p>Computes a loss value using one of several loss functions.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
                  (<code>                 </code>)
              –
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </li>
            <li>
              <b><code>expression_matrix</code></b>
                  (<code> </code>)
              –
              <div class="doc-md-description">
                <p>(n_genes x T_use) measured gene expression values.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code>        </code>)
              –
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein_matrix</code></b>
                  (<code> </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </li>
            <li>
              <b><code>psite_tensor</code></b>
                  (<code>      </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code>             </code>)
              –
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code>             </code>)
              –
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </li>
            <li>
              <b><code>n_genes,</code></b>
                  (<code><span title="n_TF">n_TF</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of genes and TF respectively.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
              –
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the β–segment) for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code>        </code>)
              –
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>loss_type</code></b>
                  (<code>         </code>)
              –
              <div class="doc-md-description">
                <p>Integer indicating the loss type (0: MSE, 1: MAE, 2: soft L1, 3: Cauchy, 4: Arctan, 5: Elastic Net, 6: Tikhonov).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>loss</code></b>              –
              <div class="doc-md-description">
                <p>The computed loss (a scalar).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.objfn.minfn.compute_predictions" class="doc doc-heading">
            <code class="highlight language-python">compute_predictions(x, regulators, tf_protein_matrix, psite_tensor, n_reg, T_use, n_genes, beta_start_indices, num_psites)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Computes the predicted expression matrix based on the decision vector x.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
                  (<code>                 </code>)
              –
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code>        </code>)
              –
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein_matrix</code></b>
                  (<code> </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </li>
            <li>
              <b><code>psite_tensor</code></b>
                  (<code>      </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code>             </code>)
              –
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code>             </code>)
              –
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </li>
            <li>
              <b><code>n_genes</code></b>
                  (<code>           </code>)
              –
              <div class="doc-md-description">
                <p>Number of genes.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
              –
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the β–segment) for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code>        </code>)
              –
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>predictions</code></b>              –
              <div class="doc-md-description">
                <p>(n_genes x T_use) predicted gene expression values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.objfn.minfn.objective_wrapper" class="doc doc-heading">
            <code class="highlight language-python">objective_wrapper(x, expression_matrix, regulators, tf_protein_matrix, psite_tensor, n_reg, T_use, n_genes, beta_start_indices, num_psites, loss_type)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Wrapper function for the objective function.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
                  (<code>                 </code>)
              –
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </li>
            <li>
              <b><code>expression_matrix</code></b>
                  (<code> </code>)
              –
              <div class="doc-md-description">
                <p>(n_genes x T_use) measured gene expression values.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code>        </code>)
              –
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein_matrix</code></b>
                  (<code> </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </li>
            <li>
              <b><code>psite_tensor</code></b>
                  (<code>      </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code>             </code>)
              –
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code>             </code>)
              –
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </li>
            <li>
              <b><code>n_genes</code></b>
                  (<code>           </code>)
              –
              <div class="doc-md-description">
                <p>Number of genes.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
              –
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the β–segment) for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code>        </code>)
              –
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>loss_type</code></b>
                  (<code>         </code>)
              –
              <div class="doc-md-description">
                <p>Integer indicating the loss type.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>loss</code></b>              –
              <div class="doc-md-description">
                <p>The computed loss (a scalar).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.opt.optrun" class="doc doc-heading">
            <code>tfopt.local.opt.optrun</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.opt.optrun.run_optimizer" class="doc doc-heading">
            <code class="highlight language-python">run_optimizer(x0, bounds, lin_cons, expression_matrix, regulators, tf_protein_matrix, psite_tensor, n_reg, T_use, n_genes, beta_start_indices, num_psites, loss_type)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Runs the optimization algorithm to minimize the objective function.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x0</code></b>
                  (<code>                 </code>)
              –
              <div class="doc-md-description">
                <p>Initial guess for the optimization variables.</p>
              </div>
            </li>
            <li>
              <b><code>bounds</code></b>
                  (<code>             </code>)
              –
              <div class="doc-md-description">
                <p>Bounds for the optimization variables.</p>
              </div>
            </li>
            <li>
              <b><code>lin_cons</code></b>
                  (<code>           </code>)
              –
              <div class="doc-md-description">
                <p>Linear constraints for the optimization problem.</p>
              </div>
            </li>
            <li>
              <b><code>expression_matrix</code></b>
                  (<code>  </code>)
              –
              <div class="doc-md-description">
                <p>(n_genes x T_use) measured gene expression values.</p>
              </div>
            </li>
            <li>
              <b><code>regulators</code></b>
                  (<code>         </code>)
              –
              <div class="doc-md-description">
                <p>(n_genes x n_reg) indices of TF regulators for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein_matrix</code></b>
                  (<code>  </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x T_use) TF protein time series.</p>
              </div>
            </li>
            <li>
              <b><code>psite_tensor</code></b>
                  (<code>       </code>)
              –
              <div class="doc-md-description">
                <p>(n_TF x n_psite_max x T_use) matrix of PSite signals (padded with zeros).</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code>              </code>)
              –
              <div class="doc-md-description">
                <p>Maximum number of regulators per gene.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code>              </code>)
              –
              <div class="doc-md-description">
                <p>Number of time points used.</p>
              </div>
            </li>
            <li>
              <b><code>n_genes,</code></b>
                  (<code><span title="n_TF">n_TF</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of genes and TF respectively.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code> </code>)
              –
              <div class="doc-md-description">
                <p>Integer array giving the starting index (in the β–segment) for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code>         </code>)
              –
              <div class="doc-md-description">
                <p>Integer array with the actual number of PSites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>loss_type</code></b>
                  (<code>          </code>)
              –
              <div class="doc-md-description">
                <p>Type of loss function to use.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    result             : Result of the optimization process, including the optimized parameters and objective value.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.optcon.construct" class="doc doc-heading">
            <code>tfopt.local.optcon.construct</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.build_fixed_arrays" class="doc doc-heading">
            <code class="highlight language-python">build_fixed_arrays(gene_ids, expression_matrix, tf_ids, tf_protein, tf_psite_data, tf_psite_labels, reg_map)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Builds fixed-shape arrays from the input data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>-</code></b>
                  (<code><span title="gene_ids">gene_ids</span></code>)
              –
              <div class="doc-md-description">
                <p>list of mRNA identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>-</code></b>
                  (<code><span title="expression_matrix">expression_matrix</span></code>)
              –
              <div class="doc-md-description">
                <p>array of shape (n_genes, T) with mRNA expression levels.</p>
              </div>
            </li>
            <li>
              <b><code>-</code></b>
                  (<code><span title="tf_ids">tf_ids</span></code>)
              –
              <div class="doc-md-description">
                <p>list of TF identifiers.</p>
              </div>
            </li>
            <li>
              <b><code>-</code></b>
                  (<code><span title="tf_protein">tf_protein</span></code>)
              –
              <div class="doc-md-description">
                <p>dict mapping TFs to their protein levels.</p>
              </div>
            </li>
            <li>
              <b><code>-</code></b>
                  (<code><span title="tf_psite_data">tf_psite_data</span></code>)
              –
              <div class="doc-md-description">
                <p>dict mapping TFs to their phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>-</code></b>
                  (<code><span title="tf_psite_labels">tf_psite_labels</span></code>)
              –
              <div class="doc-md-description">
                <p>dict mapping TFs to their phosphorylation site labels.</p>
              </div>
            </li>
            <li>
              <b><code>-</code></b>
                  (<code><span title="reg_map">reg_map</span></code>)
              –
              <div class="doc-md-description">
                <p>mapping of genes to their regulators (TFs).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>expression_matrix: array of shape (n_genes, T) with mRNA expression levels.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>regulators: array of shape (n_genes, n_reg) with TF indices.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>tf_protein_matrix: array of shape (n_TF, T) with TF protein levels.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>psite_tensor: array of shape (n_TF, n_psite_max, T) with phosphorylation sites.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>n_reg: number of regulators.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>n_psite_max: maximum number of phosphorylation sites across all TFs.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>psite_labels_arr: list of labels for each TF's phosphorylation sites.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>num_psites: array indicating the number of phosphorylation sites for each TF.</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.constraint_alpha_func" class="doc doc-heading">
            <code class="highlight language-python">constraint_alpha_func(x, n_genes, n_reg)</code>

</h3>


    <div class="doc doc-contents ">

        <p>For each gene, the sum of its alpha parameters must equal 1.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </li>
            <li>
              <b><code>n_genes</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of genes.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>np.ndarray: Array of constraints.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.constraint_beta_func" class="doc doc-heading">
            <code class="highlight language-python">constraint_beta_func(x, n_alpha, n_TF, beta_start_indices, num_psites, no_psite_tf)</code>

</h3>


    <div class="doc doc-contents ">

        <p>For each TF, the sum of its beta parameters must equal 1.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Decision vector.</p>
              </div>
            </li>
            <li>
              <b><code>n_alpha</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </li>
            <li>
              <b><code>n_TF</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of transcription factors.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of starting indices for beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>no_psite_tf</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List indicating if a TF has no phosphorylation site.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>np.ndarray: Array of constraints.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.construct.build_linear_constraints" class="doc doc-heading">
            <code class="highlight language-python">build_linear_constraints(n_genes, n_TF, n_reg, n_alpha, beta_start_indices, num_psites, no_psite_tf)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Build linear constraints for the transcription factor optimization problem.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>n_genes</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of genes.</p>
              </div>
            </li>
            <li>
              <b><code>n_TF</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of transcription factors.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>n_alpha</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of starting indices for beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>no_psite_tf</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List indicating if a TF has no phosphorylation site.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>list</code></b>              –
              <div class="doc-md-description">
                <p>List of linear constraints.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.optcon.filter" class="doc doc-heading">
            <code>tfopt.local.optcon.filter</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.filter.load_and_filter_data" class="doc doc-heading">
            <code class="highlight language-python">load_and_filter_data()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load and filter data for the optimization problem.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>gene_ids (list): List of gene IDs.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>expr_matrix (np.ndarray): Gene expression matrix.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>expr_time_cols (list): Time columns for expression data.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>tf_ids (list): List of transcription factor IDs.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>tf_protein (dict): Dictionary mapping TF IDs to their protein data.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>tf_psite_data (dict): Dictionary mapping TF IDs to their phosphorylation site data.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>tf_psite_labels (dict): Dictionary mapping TF IDs to their phosphorylation site labels.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>tf_time_cols (list): Time columns for TF data.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>reg_map (dict): Regulation map, mapping gene IDs to their regulators.</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.optcon.filter.prepare_data" class="doc doc-heading">
            <code class="highlight language-python">prepare_data(gene_ids, expr_matrix, tf_ids, tf_protein, tf_psite_data, tf_psite_labels, tf_time_cols, reg_map)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Prepares the data for optimization by filtering the expression matrix
to match the number of time points and building fixed arrays.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of gene IDs.</p>
              </div>
            </li>
            <li>
              <b><code>expr_matrix</code></b>
                  (<code><span title="np.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene expression matrix.</p>
              </div>
            </li>
            <li>
              <b><code>tf_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of transcription factor IDs.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their protein data.</p>
              </div>
            </li>
            <li>
              <b><code>tf_psite_data</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their phosphorylation site data.</p>
              </div>
            </li>
            <li>
              <b><code>tf_psite_labels</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary mapping TF IDs to their phosphorylation site labels.</p>
              </div>
            </li>
            <li>
              <b><code>tf_time_cols</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Time columns for TF data.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Regulation map, mapping gene IDs to their regulators.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    fixed_arrays (tuple): Tuple containing the fixed arrays:
        - expression_matrix: array of shape (n_genes, T)
        - regulators: array of shape (n_genes, n_reg) with indices into tf_ids.
        - tf_protein_matrix: array of shape (n_TF, T)
        - psite_tensor: array of shape (n_TF, n_psite_max, T), padded with zeros.
        - n_reg: maximum number of regulators per gene.
        - n_psite_max: maximum number of PSites among TFs.
        - psite_labels_arr: list (length n_TF) of lists of PSite names (padded with empty strings).
        - num_psites: array of length n_TF with the actual number of PSites for each TF.
    T_use (int): Number of time points used in the expression matrix.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.utils.iodata" class="doc doc-heading">
            <code>tfopt.local.utils.iodata</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.min_max_normalize" class="doc doc-heading">
            <code class="highlight language-python">min_max_normalize(df, custom_max=None)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Row-wise (per-sample) min-max normalize time-series columns starting with 'x'.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>df</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>Input DataFrame with time-series columns (x1-xN).</p>
              </div>
            </li>
            <li>
              <b><code>custom_max</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>If given, used as max for all rows.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: Normalized DataFrame with same shape.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.load_expression_data" class="doc doc-heading">
            <code class="highlight language-python">load_expression_data(filename=INPUT3)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Loads gene expression (mRNA) data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.INPUT3">INPUT3</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the CSV file containing mRNA data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>gene_ids: List of gene identifiers (strings).</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>expression_matrix: Matrix of gene expression data (numpy array).</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>time_cols: List of time columns (excluding "GeneID").</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.load_tf_protein_data" class="doc doc-heading">
            <code class="highlight language-python">load_tf_protein_data(filename=INPUT1)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Loads TF protein data along with PSite information.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.INPUT1">INPUT1</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the CSV file containing TF protein data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    - tf_ids: List of TF identifiers (strings).
    - tf_protein: Dictionary mapping TF identifiers to their protein data (numpy array).
    - tf_psite_data: Dictionary mapping TF identifiers to their phosphorylation site data (list of numpy arrays).
    - tf_psite_labels: Dictionary mapping TF identifiers to their phosphorylation site labels (list of strings).
    - time_cols: List of time columns (excluding "GeneID" and "Psite").</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.load_regulation" class="doc doc-heading">
            <code class="highlight language-python">load_regulation(filename=INPUT4)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Returns a mapping from gene (source) to a list of TFs (targets).</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.INPUT4">INPUT4</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the CSV file containing regulation data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    - reg_map: Dictionary mapping gene identifiers to lists of TF identifiers.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.summarize_stats" class="doc doc-heading">
            <code class="highlight language-python">summarize_stats(input3=INPUT3, input1=INPUT1, input4=INPUT4)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Summarizes statistics for the expression data (input3) and TF protein data (input1).</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>input3</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.INPUT3">INPUT3</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the expression data CSV file.</p>
              </div>
            </li>
            <li>
              <b><code>input1</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.INPUT1">INPUT1</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the TF protein data CSV file.</p>
              </div>
            </li>
            <li>
              <b><code>input4</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="tfopt.local.config.constants.INPUT4">INPUT4</span></code>
)
              –
              <div class="doc-md-description">
                <p>Path to the mapping file CSV.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.create_report" class="doc doc-heading">
            <code class="highlight language-python">create_report(results_dir: str, output_file: str = 'report.html')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>results_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the root results directory.</p>
              </div>
            </li>
            <li>
              <b><code>output_file</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;report.html&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.iodata.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python">organize_output_files(*directories)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to organize output files into protein-specific folders.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>directories</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>()</code>
)
              –
              <div class="doc-md-description">
                <p>List of directories to organize.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="tfopt.local.utils.params" class="doc doc-heading">
            <code>tfopt.local.utils.params</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.params.get_optimization_parameters" class="doc doc-heading">
            <code class="highlight language-python">get_optimization_parameters(expression_matrix, tf_protein_matrix, n_reg, T_use, psite_labels_arr, num_psites, lb, ub)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Prepare the optimization parameters for the optimization problem.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>expression_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene expression matrix.</p>
              </div>
            </li>
            <li>
              <b><code>tf_protein_matrix</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>TF protein matrix.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>T_use</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of time points to use.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_arr</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of phosphorylation site labels for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array containing the number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>lb</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Lower bound for beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>ub</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Upper bound for beta parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    x0 (np.ndarray): Initial guess for the optimization variables.
    n_alpha (int): Number of alpha parameters.
    beta_start_indices (np.ndarray): Starting indices for beta parameters.
    bounds (list): List of bounds for the optimization variables.
    no_psite_tf (np.ndarray): Array indicating whether each TF has no phosphorylation sites.
    n_genes (int): Number of genes.
    n_TF (int): Number of transcription factors.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="tfopt.local.utils.params.postprocess_results" class="doc doc-heading">
            <code class="highlight language-python">postprocess_results(result, n_alpha, n_genes, n_reg, beta_start_indices, num_psites, reg_map, gene_ids, tf_ids, psite_labels_arr)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Post-process the optimization results to extract the final alpha and beta parameters.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>result</code></b>
                  (<code><span title="OptimizeResult">OptimizeResult</span></code>)
              –
              <div class="doc-md-description">
                <p>The result of the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>n_alpha</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of alpha parameters.</p>
              </div>
            </li>
            <li>
              <b><code>n_genes</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of genes.</p>
              </div>
            </li>
            <li>
              <b><code>n_reg</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of regulators.</p>
              </div>
            </li>
            <li>
              <b><code>beta_start_indices</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Starting indices for beta parameters.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array containing the number of phosphorylation sites for each TF.</p>
              </div>
            </li>
            <li>
              <b><code>reg_map</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Regulation map, mapping gene IDs to their regulators.</p>
              </div>
            </li>
            <li>
              <b><code>gene_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of gene IDs.</p>
              </div>
            </li>
            <li>
              <b><code>tf_ids</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of transcription factor IDs.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels_arr</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of lists containing phosphorylation site labels.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>final_x</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Final optimization result.</p>
              </div>
            </li>
            <li>
<b><code>final_alpha</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Final alpha parameters reshaped into a matrix.</p>
              </div>
            </li>
            <li>
<b><code>final_beta</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Final beta parameters reshaped into a matrix.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h2 id="fitting-analysis">Fitting Analysis</h2>


<div class="doc doc-object doc-module">



<h2 id="tfopt.fitanalysis.helper" class="doc doc-heading">
            <code>tfopt.fitanalysis.helper</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="tfopt.fitanalysis.helper.Plotter" class="doc doc-heading">
            <code>Plotter</code>


</h3>


    <div class="doc doc-contents ">


        <p>A class to plot various analysis results from an Excel file.</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.__init__" class="doc doc-heading">
            <code class="highlight language-python">__init__(filepath, savepath)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Initializes the Plotter instance by loading data from the Excel file.
Args:
    filepath (str): Path to the Excel file containing analysis results.
    savepath (str): Directory where the plots will be saved.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.load_data" class="doc doc-heading">
            <code class="highlight language-python">load_data()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Loads data from the specified Excel file.
Args:
    filepath (str): Path to the Excel file.
    savepath (str): Directory where the plots will be saved.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_alpha_distribution" class="doc doc-heading">
            <code class="highlight language-python">plot_alpha_distribution()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the distribution of alpha parameter values grouped by transcription factors (TFs)
using a strip plot.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_beta_barplots" class="doc doc-heading">
            <code class="highlight language-python">plot_beta_barplots()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Processes the beta values DataFrame and creates a separate bar plot
for each unique transcription factor (TF).</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_heatmap_abs_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_heatmap_abs_residuals()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots a heatmap of the absolute values of the residuals.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_goodness_of_fit" class="doc doc-heading">
            <code class="highlight language-python">plot_goodness_of_fit()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Creates a scatter plot comparing observed vs. estimated values,
fits a linear regression model, plots the 95% confidence interval,
and labels points outside the confidence interval.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_kld" class="doc doc-heading">
            <code class="highlight language-python">plot_kld()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the Kullback-Leibler Divergence (KLD) for each mRNA.
The KLD is calculated between the observed and estimated distributions
of the mRNA expression levels.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_pca" class="doc doc-heading">
            <code class="highlight language-python">plot_pca()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots a PCA (Principal Component Analysis) of the observed and estimated values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_boxplot_alpha" class="doc doc-heading">
            <code class="highlight language-python">plot_boxplot_alpha()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots a boxplot of the alpha values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_boxplot_beta" class="doc doc-heading">
            <code class="highlight language-python">plot_boxplot_beta()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots a boxplot of the beta values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_cdf_alpha" class="doc doc-heading">
            <code class="highlight language-python">plot_cdf_alpha()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the cumulative distribution function (CDF) of the alpha values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_cdf_beta" class="doc doc-heading">
            <code class="highlight language-python">plot_cdf_beta()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the cumulative distribution function (CDF) of the beta values.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="tfopt.fitanalysis.helper.Plotter.plot_time_wise_residuals" class="doc doc-heading">
            <code class="highlight language-python">plot_time_wise_residuals()</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the residuals over time for each mRNA.</p>


    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h2 id="ode-modelling-parameter-estimation">ODE Modelling &amp; Parameter Estimation</h2>
<h3 id="configuration">Configuration</h3>


<div class="doc doc-object doc-module">



<h2 id="config.cli" class="doc doc-heading">
            <code>config.cli</code>


</h2>

    <div class="doc doc-contents first">

        <p>Command‑line entry point for the phoskintime pipeline.</p>
<h4 id="config.cli--usage">Usage</h4>
<h3 id="config.cli--come-one-level-up-from-the-package-root-it-should-be-the-working-directory">Come one level up from the package root, it should be the working directory</h3>
<h3 id="config.cli--where-you-can-see-the-project-directory">(where you can see the project directory).</h3>
<h3 id="config.cli--run-everything-with-the-default-local-solver">run everything with the default (local) solver</h3>
<p>python phoskintime all</p>
<h3 id="config.cli--run-only-preprocessing">run only preprocessing</h3>
<p>python phoskintime prep</p>
<h3 id="config.cli--run-tfopt-with-local-flavour">run tfopt with local flavour</h3>
<p>python phoskintime tfopt --mode local</p>
<h3 id="config.cli--run-tfopt-with-evol-flavour">run tfopt with evol flavour</h3>
<p>python phoskintime tfopt --mode evol</p>
<h3 id="config.cli--run-kinopt-with-local-flavour">run kinopt with local flavour</h3>
<p>python phoskintime kinopt --mode local</p>
<h3 id="config.cli--run-kinopt-with-evol-flavour">run kinopt with evol flavour</h3>
<p>python phoskintime kinopt --mode evol</p>
<h3 id="config.cli--run-the-model">run the model</h3>
<p>python phoskintime model</p>









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="config.cli.prep" class="doc doc-heading">
            <code class="highlight language-python">prep()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Preprocess data (processing.cleanup).</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.cli.tfopt" class="doc doc-heading">
            <code class="highlight language-python">tfopt(mode: str = typer.Option('local', help='local | evol'), conf: Path | None = typer.Option(None, '--conf', file_okay=True, dir_okay=False, writable=False, help='Path to TOML/YAML config. Uses defaults if omitted.'))</code>

</h3>


    <div class="doc doc-contents ">

        <p>Transcription-Factor-mRNA Optimisation.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mode</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="typer.Option">Option</span>(&#39;local&#39;, help=&#39;local | evol&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>local | evol</p>
              </div>
            </li>
            <li>
              <b><code>conf</code></b>
                  (<code><span title="pathlib.Path">Path</span> | None</code>, default:
                      <code><span title="typer.Option">Option</span>(None, &#39;--conf&#39;, file_okay=True, dir_okay=False, writable=False, help=&#39;Path to TOML/YAML config. Uses defaults if omitted.&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>Path to TOML/YAML config. Uses defaults if omitted.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    None</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.cli.kinopt" class="doc doc-heading">
            <code class="highlight language-python">kinopt(mode: str = typer.Option('local', help='local | evol'), conf: Path | None = typer.Option(None, '--conf', file_okay=True, dir_okay=False, writable=False, help='Path to TOML/YAML config. Uses defaults if omitted.'))</code>

</h3>


    <div class="doc doc-contents ">

        <p>Kinase-Phosphorylation Optimization.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>mode</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="typer.Option">Option</span>(&#39;local&#39;, help=&#39;local | evol&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>local | evol</p>
              </div>
            </li>
            <li>
              <b><code>conf</code></b>
                  (<code><span title="pathlib.Path">Path</span> | None</code>, default:
                      <code><span title="typer.Option">Option</span>(None, &#39;--conf&#39;, file_okay=True, dir_okay=False, writable=False, help=&#39;Path to TOML/YAML config. Uses defaults if omitted.&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>Path to TOML/YAML config. Uses defaults if omitted.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    None</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.cli.model" class="doc doc-heading">
            <code class="highlight language-python">model(conf: Path | None = typer.Option(None, '--conf', file_okay=True, dir_okay=False, writable=False, help='Path to model config file. Uses defaults if omitted.'))</code>

</h3>


    <div class="doc doc-contents ">

        <p>Run the model (bin.main).</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>conf</code></b>
                  (<code><span title="pathlib.Path">Path</span> | None</code>, default:
                      <code><span title="typer.Option">Option</span>(None, &#39;--conf&#39;, file_okay=True, dir_okay=False, writable=False, help=&#39;Path to model config file. Uses defaults if omitted.&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>Path to model config file. Uses defaults if omitted.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    None</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.cli.all" class="doc doc-heading">
            <code class="highlight language-python">all(tf_mode: str = typer.Option('local', help='tfopt mode: local | evol'), kin_mode: str = typer.Option('local', help='kinopt mode: local | evol'), tf_conf: Path | None = typer.Option(None, help='tfopt config file'), kin_conf: Path | None = typer.Option(None, help='kinopt config file'), model_conf: Path | None = typer.Option(None, help='model config file'))</code>

</h3>


    <div class="doc doc-contents ">

        <p>Run every stage in sequence.
Preprocessing -&gt; TF optimisation -&gt; Kinase optimisation -&gt; Model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>tf_mode</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="typer.Option">Option</span>(&#39;local&#39;, help=&#39;tfopt mode: local | evol&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>tfopt mode: local | evol</p>
              </div>
            </li>
            <li>
              <b><code>kin_mode</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="typer.Option">Option</span>(&#39;local&#39;, help=&#39;kinopt mode: local | evol&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>kinopt mode: local | evol</p>
              </div>
            </li>
            <li>
              <b><code>tf_conf</code></b>
                  (<code><span title="pathlib.Path">Path</span> | None</code>, default:
                      <code><span title="typer.Option">Option</span>(None, help=&#39;tfopt config file&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>Path to TOML/YAML config. Uses defaults if omitted.</p>
              </div>
            </li>
            <li>
              <b><code>kin_conf</code></b>
                  (<code><span title="pathlib.Path">Path</span> | None</code>, default:
                      <code><span title="typer.Option">Option</span>(None, help=&#39;kinopt config file&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>Path to TOML/YAML config. Uses defaults if omitted.</p>
              </div>
            </li>
            <li>
              <b><code>model_conf</code></b>
                  (<code><span title="pathlib.Path">Path</span> | None</code>, default:
                      <code><span title="typer.Option">Option</span>(None, help=&#39;model config file&#39;)</code>
)
              –
              <div class="doc-md-description">
                <p>Path to model config file. Uses defaults if omitted.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    None</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="config.config" class="doc doc-heading">
            <code>config.config</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="config.config.parse_bound_pair" class="doc doc-heading">
            <code class="highlight language-python">parse_bound_pair(val)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Parse a string representing a pair of bounds (lower, upper) into a tuple of floats.
The upper bound can be 'inf' or 'infinity' to represent infinity.
Raises ValueError if the input is not in the correct format.
Args:
    val (str): The string to parse, e.g., "0,3" or "0,infinity".
Returns:
    tuple: A tuple containing the lower and upper bounds as floats.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.parse_fix_value" class="doc doc-heading">
            <code class="highlight language-python">parse_fix_value(val)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Parse a fixed value or a list of fixed values from a string.
If the input is a single value, it returns that value as a float.
If the input is a comma-separated list, it returns a list of floats.
Raises ValueError if the input is not in the correct format.
Args:
    val (str): The string to parse, e.g., "1.0" or "1.0,2.0".
Returns:
    float or list: The parsed fixed value(s) as a float or a list of floats.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.ensure_output_directory" class="doc doc-heading">
            <code class="highlight language-python">ensure_output_directory(directory)</code>

</h3>


    <div class="doc doc-contents ">



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>directory</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The path to the directory to create.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    None</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.parse_args" class="doc doc-heading">
            <code class="highlight language-python">parse_args()</code>

</h3>


    <div class="doc doc-contents ">

        <p>Parse command-line arguments for the PhosKinTime script.
This function uses argparse to define and handle the command-line options.
It includes options for setting bounds, fixed parameters, bootstrapping,
profile estimation, and input file paths.
The function returns the parsed arguments as a Namespace object.
The arguments include:
--A-bound, --B-bound, --C-bound, --D-bound,
--Ssite-bound, --Dsite-bound, --bootstraps,
--input-excel-protein, --input-excel-psite, --input-excel-rna.</p>
        <p>Returns:
    argparse.Namespace: The parsed command-line arguments.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.log_config" class="doc doc-heading">
            <code class="highlight language-python">log_config(logger, bounds, args)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Log the configuration settings for the PhosKinTime script.
This function logs the parameter bounds
bootstrapping iterations.
It uses the provided logger to output the information.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>logger</code></b>
                  (<code><span title="config.config.logging.Logger">Logger</span></code>)
              –
              <div class="doc-md-description">
                <p>The logger to use for logging.</p>
              </div>
            </li>
            <li>
              <b><code>bounds</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>The parameter bounds.</p>
              </div>
            </li>
            <li>
              <b><code>args</code></b>
                  (<code><span title="argparse.Namespace">Namespace</span></code>)
              –
              <div class="doc-md-description">
                <p>The command-line arguments.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    None</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.extract_config" class="doc doc-heading">
            <code class="highlight language-python">extract_config(args)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Extract configuration settings from command-line arguments.
This function creates a dictionary containing the parameter bounds, bootstrapping iterations.
The function returns the configuration dictionary.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>args</code></b>
                  (<code><span title="argparse.Namespace">Namespace</span></code>)
              –
              <div class="doc-md-description">
                <p>The command-line arguments.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    dict: The configuration settings.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.score_fit" class="doc doc-heading">
            <code class="highlight language-python">score_fit(params, target, prediction, alpha=ALPHA_WEIGHT, beta=BETA_WEIGHT, gamma=GAMMA_WEIGHT, delta=DELTA_WEIGHT, mu=MU_WEIGHT)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Calculate the score for the fit of a model to target data.
The score is a weighted combination of various metrics including
mean squared error (MSE), root mean squared error (RMSE),
mean absolute error (MAE), variance, and regularization penalty.
The weights for each metric can be adjusted using the parameters
alpha, beta, gamma, and delta.
The regularization penalty is controlled by the reg_penalty parameter.
The function returns the calculated score.
Args:
    params (np.ndarray): The model parameters.
    target (np.ndarray): The target data.
    prediction (np.ndarray): The predicted data.
    alpha (float): Weight for RMSE.
    beta (float): Weight for MAE.
    gamma (float): Weight for variance.
    delta (float): Weight for MSE.
    mu (float): Regularization penalty weight.
Returns:
    float: The calculated score.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.config.future_times" class="doc doc-heading">
            <code class="highlight language-python">future_times(n_new: int, ratio: Optional[float] = None, tp: np.ndarray = TIME_POINTS) -&gt; np.ndarray</code>

</h3>


    <div class="doc doc-contents ">

        <p>Extend ttime points by n_new points, each spaced by multiplying the previous interval by ratio.
If ratio is None, it is inferred from the last two points.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>n_new</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of new time points to generate.</p>
              </div>
            </li>
            <li>
              <b><code>ratio</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Ratio to multiply the previous interval. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>tp</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>, default:
                      <code><span title="config.constants.TIME_POINTS">TIME_POINTS</span></code>
)
              –
              <div class="doc-md-description">
                <p>Existing time points. Defaults to TIME_POINTS.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    np.ndarray: Extended time points.</p>


    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="config.constants" class="doc doc-heading">
            <code>config.constants</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="config.constants.get_param_names_rand" class="doc doc-heading">
            <code class="highlight language-python">get_param_names_rand(num_psites: int) -&gt; list</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generate parameter names for the random model.
Format: ['A', 'B', 'C', 'D'] +
        ['S1', 'S2', ..., 'S<num_psites>'] +
        [parameter names for all combinations of dephosphorylation sites].</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    list: List of parameter names.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.get_param_names_ds" class="doc doc-heading">
            <code class="highlight language-python">get_param_names_ds(num_psites: int) -&gt; list</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generate parameter names for distributive or successive models.
Format: ['A', 'B', 'C', 'D'] +
        ['S1', 'S2', ..., 'S<num_psites>'] +
        ['D1', 'D2', ..., 'D<num_psites>'].</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    list: List of parameter names.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.generate_labels_rand" class="doc doc-heading">
            <code class="highlight language-python">generate_labels_rand(num_psites: int) -&gt; list</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generates labels for the states based on the number of phosphorylation sites for the random model.
Returns a list with the base labels "R" and "P", followed by labels for all combinations of phosphorylated sites.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    list: List of state labels.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.generate_labels_ds" class="doc doc-heading">
            <code class="highlight language-python">generate_labels_ds(num_psites: int) -&gt; list</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generates labels for the states based on the number of phosphorylation sites for the distributive or successive models.
Returns a list with the base labels "R" and "P", followed by labels for each individual phosphorylated state.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    list: List of state labels.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.location" class="doc doc-heading">
            <code class="highlight language-python">location(path: str, label: str = None) -&gt; str</code>

</h3>


    <div class="doc doc-contents ">

        <p>Returns a clickable hyperlink string for supported terminals using ANSI escape sequences.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The file path or URL.</p>
              </div>
            </li>
            <li>
              <b><code>label</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The display text for the link. Defaults to the path if not provided.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>str</code></b>(                  <code><span title="str">str</span></code>
)              –
              <div class="doc-md-description">
                <p>A string that, when printed, shows a clickable link in terminals that support ANSI hyperlinks.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.get_number_of_params_rand" class="doc doc-heading">
            <code class="highlight language-python">get_number_of_params_rand(num_psites)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Calculate the number of parameters required for the ODE system based on the number of phosphorylation sites.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites (1 to 4).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>int</code></b>              –
              <div class="doc-md-description">
                <p>Total number of parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="config.constants.get_bounds_rand" class="doc doc-heading">
            <code class="highlight language-python">get_bounds_rand(num_psites, ub=0, lower=0)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generate bounds for the ODE parameters based on the number of phosphorylation sites.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>lower</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>0</code>
)
              –
              <div class="doc-md-description">
                <p>Lower bound for parameters.</p>
              </div>
            </li>
            <li>
              <b><code>upper</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Upper bound for parameters.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>list</code></b>              –
              <div class="doc-md-description">
                <p>List of bounds as [lower, upper] for each parameter.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="config.logconf" class="doc doc-heading">
            <code>config.logconf</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="config.logconf.ColoredFormatter" class="doc doc-heading">
            <code>ColoredFormatter</code>


</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="logging.Formatter">Formatter</span></code></p>


        <p>Custom formatter to add colors to log messages and elapsed time.
This formatter uses ANSI escape codes to colorize the log messages based on their severity level.
It also includes a right-aligned clock that shows the elapsed time since the logger was initialized.
The elapsed time is displayed in a human-readable format (e.g., "1h 23m 45s").
The formatter is designed to be used with a logger that has a console handler.
The elapsed time is calculated from the time the logger was initialized and is displayed in a right-aligned format.
The formatter also ensures that the log messages are padded to a specified width, which can be adjusted using the <code>width</code> parameter.
The <code>remove_ansi</code> method is used to strip ANSI escape codes from the log message for accurate padding calculation.
The <code>format</code> method is overridden to customize the log message format, including the timestamp, logger name, log level, and message.
The <code>setup_logger</code> function is used to configure the logger with a file handler and a stream handler.
The file handler writes log messages to a specified log file, while the stream handler outputs log messages to the console.
The logger is set to the specified logging level, and the log file is created in the specified directory.
The log file is rotated based on size, and old log files are backed up.</p>










  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="config.logconf.ColoredFormatter.format" class="doc doc-heading">
            <code class="highlight language-python">format(record)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Format the log record with colors and elapsed time.
This method overrides the default format method to customize the log message format.
It includes the timestamp, logger name, log level, and message.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="config.logconf.ColoredFormatter.remove_ansi" class="doc doc-heading">
            <code class="highlight language-python">remove_ansi(s)</code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

        <p>Remove ANSI escape codes from a string.</p>


    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="config.logconf.setup_logger" class="doc doc-heading">
            <code class="highlight language-python">setup_logger(name='phoskintime', log_file=None, level=logging.DEBUG, log_dir=LOG_DIR, rotate=True, max_bytes=2 * 1024 * 1024, backup_count=5)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Setup a logger with colored output and file logging.
This function creates a logger with colored output for console messages
:param name:
:param log_file:
:param level:
:param log_dir:
:param rotate:
:param max_bytes:
:param backup_count:
:return: logger</p>


    </div>

</div>



  </div>

    </div>

</div><h3 id="core-functions">Core Functions</h3>


<div class="doc doc-object doc-module">



<h2 id="paramest.normest" class="doc doc-heading">
            <code>paramest.normest</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.normest.worker_find_lambda" class="doc doc-heading">
            <code class="highlight language-python">worker_find_lambda(lam: float, gene: str, target: np.ndarray, p0: np.ndarray, time_points: np.ndarray, free_bounds: Tuple[np.ndarray, np.ndarray], init_cond: np.ndarray, num_psites: int, p_data: np.ndarray, pr_data: np.ndarray) -&gt; Tuple[float, float, str]</code>

</h3>


    <div class="doc doc-contents ">

        <p>Worker function for a single lambda value.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>lam</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Regularization parameter.</p>
              </div>
            </li>
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene name.</p>
              </div>
            </li>
            <li>
              <b><code>target</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Target data.</p>
              </div>
            </li>
            <li>
              <b><code>p0</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Initial parameter guess.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points for the model fitting.</p>
              </div>
            </li>
            <li>
              <b><code>free_bounds</code></b>
                  (<code><span title="typing.Tuple">Tuple</span>[<span title="numpy.ndarray">ndarray</span>, <span title="numpy.ndarray">ndarray</span>]</code>)
              –
              <div class="doc-md-description">
                <p>Parameter bounds for the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>init_cond</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Initial conditions for the ODE solver.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>p_data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Measurement data for protein-phospho.</p>
              </div>
            </li>
            <li>
              <b><code>pr_data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Reference data for protein.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="typing.Tuple">Tuple</span>[<span title="float">float</span>, <span title="float">float</span>, <span title="str">str</span>]</code>
              –
              <div class="doc-md-description">
                <p>Tuple containing the lambda value, score, and weight key.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.normest.find_best_lambda" class="doc doc-heading">
            <code class="highlight language-python">find_best_lambda(gene: str, target: np.ndarray, p0: np.ndarray, time_points: np.ndarray, free_bounds: Tuple[np.ndarray, np.ndarray], init_cond: np.ndarray, num_psites: int, p_data: np.ndarray, pr_data: np.ndarray, lambdas=np.logspace(-2, 0, 10), max_workers: int = os.cpu_count()) -&gt; Tuple[float, str]</code>

</h3>


    <div class="doc doc-contents ">

        <p>Finds best lambda_reg to use in model_func.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.normest.normest" class="doc doc-heading">
            <code class="highlight language-python">normest(gene, pr_data, p_data, r_data, init_cond, num_psites, time_points, bounds, bootstraps, use_regularization=USE_REGULARIZATION)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to estimate parameters for a given gene using ODE models.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
              –
              <div class="doc-md-description">
                <p>Gene name.</p>
              </div>
            </li>
            <li>
              <b><code>pr_data</code></b>
              –
              <div class="doc-md-description">
                <p>Protein data.</p>
              </div>
            </li>
            <li>
              <b><code>p_data</code></b>
              –
              <div class="doc-md-description">
                <p>Phosphorylation data.</p>
              </div>
            </li>
            <li>
              <b><code>r_data</code></b>
              –
              <div class="doc-md-description">
                <p>Reference data.</p>
              </div>
            </li>
            <li>
              <b><code>init_cond</code></b>
              –
              <div class="doc-md-description">
                <p>Initial conditions for the ODE solver.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
              –
              <div class="doc-md-description">
                <p>Time points for the model fitting.</p>
              </div>
            </li>
            <li>
              <b><code>bounds</code></b>
              –
              <div class="doc-md-description">
                <p>Parameter bounds for the optimization.</p>
              </div>
            </li>
            <li>
              <b><code>bootstraps</code></b>
              –
              <div class="doc-md-description">
                <p>Number of bootstrap iterations.</p>
              </div>
            </li>
            <li>
              <b><code>use_regularization</code></b>
              –
              <div class="doc-md-description">
                <p>Whether to use regularization in the fitting process.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>Tuple containing estimated parameters, model fits, error values, and regularization term.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="paramest.toggle" class="doc doc-heading">
            <code>paramest.toggle</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.toggle.estimate_parameters" class="doc doc-heading">
            <code class="highlight language-python">estimate_parameters(gene, pr_data, p_data, r_data, init_cond, num_psites, time_points, bounds, bootstraps)</code>

</h3>


    <div class="doc doc-contents ">

        <p>This function allows for the selection of the estimation mode
and handles the parameter estimation process accordingly.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene name.</p>
              </div>
            </li>
            <li>
              <b><code>pr_data</code></b>
                  (<code><span title="array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of protein data.</p>
              </div>
            </li>
            <li>
              <b><code>p_data</code></b>
                  (<code><span title="array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of protein-phospho data.</p>
              </div>
            </li>
            <li>
              <b><code>r_data</code></b>
                  (<code><span title="array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of RNA data.</p>
              </div>
            </li>
            <li>
              <b><code>init_cond</code></b>
                  (<code><span title="array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Initial conditions for the model.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points for the data.</p>
              </div>
            </li>
            <li>
              <b><code>bounds</code></b>
                  (<code><span title="tuple">tuple</span></code>)
              –
              <div class="doc-md-description">
                <p>Bounds for the parameter estimation.</p>
              </div>
            </li>
            <li>
              <b><code>bootstraps</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of bootstrap samples.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>model_fits</code></b>(                  <code><span title="list">list</span></code>
)              –
              <div class="doc-md-description">
                <p>List of model fits.</p>
              </div>
            </li>
            <li>
<b><code>estimated_params</code></b>(                  <code><span title="array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Estimated parameters.</p>
              </div>
            </li>
            <li>
<b><code>seq_model_fit</code></b>(                  <code><span title="array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Sequence model fit.</p>
              </div>
            </li>
            <li>
<b><code>errors</code></b>(                  <code><span title="array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Errors in the estimation.</p>
              </div>
            </li>
            <li>
<b><code>reg_term</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>Regularization term.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="weights-for-curve-fitting">Weights for Curve Fitting</h3>


<div class="doc doc-object doc-module">



<h2 id="models.weights" class="doc doc-heading">
            <code>models.weights</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.weights.early_emphasis" class="doc doc-heading">
            <code class="highlight language-python">early_emphasis(pr_data, p_data, time_points, num_psites)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function that calculates custom weights for early time points in a dataset.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>pr_data</code></b>
              –
              <div class="doc-md-description">
                <p>2D numpy array of shape (num_psites, n_times)</p>
              </div>
            </li>
            <li>
              <b><code>p_data</code></b>
              –
              <div class="doc-md-description">
                <p>2D numpy array of shape (num_psites, n_times)</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
              –
              <div class="doc-md-description">
                <p>1D numpy array of time points</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>custom_weights</code></b>              –
              <div class="doc-md-description">
                <p>1D numpy array of weights for early time points</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.weights.get_protein_weights" class="doc doc-heading">
            <code class="highlight language-python">get_protein_weights(gene, input1_path=Path(__file__).resolve().parent.parent / 'processing' / 'input1_wstd.csv', input2_path=Path(__file__).resolve().parent.parent / 'kinopt' / 'data' / 'input2.csv')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to extract weights for a specific gene from the input files.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene ID to filter the weights.</p>
              </div>
            </li>
            <li>
              <b><code>input1_path</code></b>
                  (<code><span title="pathlib.Path">Path</span></code>, default:
                      <code><span title="parent.parent">parent</span> / &#39;processing&#39; / &#39;input1_wstd.csv&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Path to the input1_wstd.csv file.</p>
              </div>
            </li>
            <li>
              <b><code>input2_path</code></b>
                  (<code><span title="pathlib.Path">Path</span></code>, default:
                      <code><span title="parent.parent">parent</span> / &#39;kinopt&#39; / &#39;data&#39; / &#39;input2.csv&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Path to the input2.csv file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>weights</code></b>(                  <code><span title="numpy.ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Extracted weights for the specified gene.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.weights.full_weight" class="doc doc-heading">
            <code class="highlight language-python">full_weight(p_data_weight, use_regularization, reg_len)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to create a full weight array for parameter estimation.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>p_data_weight</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The weight data to be processed.</p>
              </div>
            </li>
            <li>
              <b><code>use_regularization</code></b>
                  (<code><span title="bool">bool</span></code>)
              –
              <div class="doc-md-description">
                <p>Flag to indicate if regularization is used.</p>
              </div>
            </li>
            <li>
              <b><code>reg_len</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Length of the regularization term.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>numpy.ndarray: The full weight array.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.weights.get_weight_options" class="doc doc-heading">
            <code class="highlight language-python">get_weight_options(target, t_target, num_psites, use_regularization, reg_len, early_weights, ms_gauss_weights)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to calculate weights for parameter estimation based on the target data and time points.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>target</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The target data for which weights are calculated.</p>
              </div>
            </li>
            <li>
              <b><code>t_target</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The time points corresponding to the target data.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>use_regularization</code></b>
                  (<code><span title="bool">bool</span></code>)
              –
              <div class="doc-md-description">
                <p>Flag to indicate if regularization is used.</p>
              </div>
            </li>
            <li>
              <b><code>reg_len</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Length of the regularization term.</p>
              </div>
            </li>
            <li>
              <b><code>early_weights</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Weights for early time points.</p>
              </div>
            </li>
            <li>
              <b><code>ms_gauss_weights</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Weights based on Gaussian distribution.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary containing different weight options.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="parameter-estimation">Parameter Estimation</h3>


<div class="doc doc-object doc-module">



<h2 id="paramest.core" class="doc doc-heading">
            <code>paramest.core</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.core.process_gene" class="doc doc-heading">
            <code class="highlight language-python">process_gene(gene, protein_data, kinase_data, mrna_data, time_points, bounds, bootstraps=0, out_dir=OUT_DIR)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Process a single gene by estimating its parameters and generating plots.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene name.</p>
              </div>
            </li>
            <li>
              <b><code>protein_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing protein-only data.</p>
              </div>
            </li>
            <li>
              <b><code>kinase_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing kinase data.</p>
              </div>
            </li>
            <li>
              <b><code>mrna_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing mRNA data.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of time points for the experiment.</p>
              </div>
            </li>
            <li>
              <b><code>bounds</code></b>
                  (<code><span title="tuple">tuple</span></code>)
              –
              <div class="doc-md-description">
                <p>Bounds for parameter estimation.</p>
              </div>
            </li>
            <li>
              <b><code>bootstraps</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>0</code>
)
              –
              <div class="doc-md-description">
                <p>Number of bootstrap iterations. Defaults to 0.</p>
              </div>
            </li>
            <li>
              <b><code>out_dir</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="config.constants.OUT_DIR">OUT_DIR</span></code>
)
              –
              <div class="doc-md-description">
                <p>Output directory for saving results. Defaults to OUT_DIR.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>gene: The gene being processed.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>estimated_params: Estimated parameters for the gene.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>model_fits: Model fits for the gene.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>seq_model_fit: Sequential model fit for the gene.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>errors: Error metrics (MSE, MAE).</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>final_params: Final estimated parameters.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>param_df: DataFrame of estimated parameters.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>gene_psite_data: Dictionary of gene-specific data.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>psite_labels: Labels for phosphorylation sites.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>pca_result: PCA result for the gene.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>ev: Explained variance for PCA.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>tsne_result: t-SNE result for the gene.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>perturbation_analysis: Sensitivity analysis results.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>perturbation_curves_params: Trajectories with parameters for sensitivity analysis.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>knockout_results: Dictionary of knockout results.</li>
</ul>
              </div>
            </li>
            <li>
              –
              <div class="doc-md-description">
                <ul>
<li>regularization: Regularization value used in parameter estimation.</li>
</ul>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="paramest.core.process_gene_wrapper" class="doc doc-heading">
            <code class="highlight language-python">process_gene_wrapper(gene, protein_data, kinase_data, mrna_data, time_points, bounds, bootstraps, out_dir=OUT_DIR)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Wrapper function to process a gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene name.</p>
              </div>
            </li>
            <li>
              <b><code>protein_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing protein-only data.</p>
              </div>
            </li>
            <li>
              <b><code>kinase_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing kinase data.</p>
              </div>
            </li>
            <li>
              <b><code>mrna_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing mRNA data.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of time points for the experiment.</p>
              </div>
            </li>
            <li>
              <b><code>bounds</code></b>
                  (<code><span title="tuple">tuple</span></code>)
              –
              <div class="doc-md-description">
                <p>Bounds for parameter estimation.</p>
              </div>
            </li>
            <li>
              <b><code>bootstraps</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of bootstrap iterations. Defaults to 0.</p>
              </div>
            </li>
            <li>
              <b><code>out_dir</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code><span title="config.constants.OUT_DIR">OUT_DIR</span></code>
)
              –
              <div class="doc-md-description">
                <p>Output directory for saving results. Defaults to OUT_DIR.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary containing the results of the gene processing.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="confidence-intervals-using-linearization">Confidence Intervals using Linearization</h3>


<div class="doc doc-object doc-module">



<h2 id="paramest.identifiability.ci" class="doc doc-heading">
            <code>paramest.identifiability.ci</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="paramest.identifiability.ci.confidence_intervals" class="doc doc-heading">
            <code class="highlight language-python">confidence_intervals(gene, popt, pcov, target, model, alpha_val=0.05)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Computes the confidence intervals for parameter estimates using Wald Intervals approach.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Gene name.</p>
              </div>
            </li>
            <li>
              <b><code>popt</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Optimized parameter estimates.</p>
              </div>
            </li>
            <li>
              <b><code>pcov</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Covariance matrix of the optimized parameters.</p>
              </div>
            </li>
            <li>
              <b><code>target</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Target data.</p>
              </div>
            </li>
            <li>
              <b><code>model</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Model predictions.</p>
              </div>
            </li>
            <li>
              <b><code>alpha_val</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>0.05</code>
)
              –
              <div class="doc-md-description">
                <p>Significance level for confidence intervals. Defaults to 0.05.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary containing the confidence intervals and other statistics.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="knockout-analysis">Knockout Analysis</h3>


<div class="doc doc-object doc-module">



<h2 id="knockout.helper" class="doc doc-heading">
            <code>knockout.helper</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">











  </div>

    </div>

</div><h3 id="perturbation-parameter-sensitivity-analysis">Perturbation &amp; Parameter Sensitivity Analysis</h3>


<div class="doc doc-object doc-module">



<h2 id="sensitivity.analysis" class="doc doc-heading">
            <code>sensitivity.analysis</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="sensitivity.analysis.compute_bound" class="doc doc-heading">
            <code class="highlight language-python">compute_bound(value, perturbation=PERTURBATIONS_VALUE)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Computes the lower and upper bounds for a given parameter value for sensitivity analysis and perturbations.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>value</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>The parameter value.</p>
              </div>
            </li>
            <li>
              <b><code>perturbation</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code><span title="config.constants.PERTURBATIONS_VALUE">PERTURBATIONS_VALUE</span></code>
)
              –
              <div class="doc-md-description">
                <p>The perturbation factor.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>list</code></b>              –
              <div class="doc-md-description">
                <p>A list containing the lower and upper bounds.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sensitivity.analysis.define_sensitivity_problem_rand" class="doc doc-heading">
            <code class="highlight language-python">define_sensitivity_problem_rand(num_psites, values)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Defines the Morris sensitivity analysis problem for the random model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>values</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of parameter values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary containing the number of variables, parameter names, and bounds.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="sensitivity.analysis.define_sensitivity_problem_ds" class="doc doc-heading">
            <code class="highlight language-python">define_sensitivity_problem_ds(num_psites, values)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Defines the Morris sensitivity analysis problem for the dynamic-site model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>values</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of parameter values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary containing the number of variables, parameter names, and bounds.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="model-diagram">Model Diagram</h3>


<div class="doc doc-object doc-module">



<h2 id="models.diagram.helpers" class="doc doc-heading">
            <code>models.diagram.helpers</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.powerset" class="doc doc-heading">
            <code class="highlight language-python">powerset(iterable)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Return the list of all subsets (as frozensets) of the given iterable.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>iterable</code></b>
              –
              <div class="doc-md-description">
                <p>An iterable (e.g., list, set) to generate subsets from.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    A list of frozensets representing all subsets of the input iterable.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.state_label" class="doc doc-heading">
            <code class="highlight language-python">state_label(state)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Convert a set of phosphorylation sites into a node label.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>state</code></b>
              –
              <div class="doc-md-description">
                <p>A frozenset representing the phosphorylation state.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    A string representing the label for the node.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.create_random_diagram" class="doc doc-heading">
            <code class="highlight language-python">create_random_diagram(x, num_sites, output_filename)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Create a random phosphorylation diagram.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
              –
              <div class="doc-md-description">
                <p>Placeholder parameter, not used in this function.</p>
              </div>
            </li>
            <li>
              <b><code>num_sites</code></b>
              –
              <div class="doc-md-description">
                <p>The number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>output_filename</code></b>
              –
              <div class="doc-md-description">
                <p>The name of the output file for the diagram.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.create_distributive_diagram" class="doc doc-heading">
            <code class="highlight language-python">create_distributive_diagram(x, num_sites, output_filename)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Create a distributive phosphorylation diagram.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
              –
              <div class="doc-md-description">
                <p>Placeholder parameter, not used in this function.</p>
              </div>
            </li>
            <li>
              <b><code>num_sites</code></b>
              –
              <div class="doc-md-description">
                <p>The number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>output_filename</code></b>
              –
              <div class="doc-md-description">
                <p>The name of the output file for the diagram.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.diagram.helpers.create_successive_model" class="doc doc-heading">
            <code class="highlight language-python">create_successive_model(x, num_sites, output_filename)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Create a successive phosphorylation diagram.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>x</code></b>
              –
              <div class="doc-md-description">
                <p>Placeholder parameter, not used in this function.</p>
              </div>
            </li>
            <li>
              <b><code>num_sites</code></b>
              –
              <div class="doc-md-description">
                <p>The number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>output_filename</code></b>
              –
              <div class="doc-md-description">
                <p>The name of the output file for the diagram.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="model-types">Model Types</h3>


<div class="doc doc-object doc-module">



<h2 id="models.distmod" class="doc doc-heading">
            <code>models.distmod</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.distmod.ode_core" class="doc doc-heading">
            <code class="highlight language-python">ode_core(y, t, A, B, C, D, S_rates, D_rates)</code>

</h3>


    <div class="doc doc-contents ">

        <p>The core ODE system for the distributive phosphorylation model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>y</code></b>
              –
              <div class="doc-md-description">
                <p>array of concentrations</p>
              </div>
            </li>
            <li>
              <b><code>t</code></b>
              –
              <div class="doc-md-description">
                <p>time</p>
              </div>
            </li>
            <li>
              <b><code>A</code></b>
              –
              <div class="doc-md-description">
                <p>mRNA production rate</p>
              </div>
            </li>
            <li>
              <b><code>B</code></b>
              –
              <div class="doc-md-description">
                <p>mRNA degradation rate</p>
              </div>
            </li>
            <li>
              <b><code>C</code></b>
              –
              <div class="doc-md-description">
                <p>protein production rate</p>
              </div>
            </li>
            <li>
              <b><code>D</code></b>
              –
              <div class="doc-md-description">
                <p>protein degradation rate</p>
              </div>
            </li>
            <li>
              <b><code>S_rates</code></b>
              –
              <div class="doc-md-description">
                <p>phosphorylation rates for each site</p>
              </div>
            </li>
            <li>
              <b><code>D_rates</code></b>
              –
              <div class="doc-md-description">
                <p>dephosphorylation rates for each site</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dydt</code></b>              –
              <div class="doc-md-description">
                <p>array of derivatives</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.distmod.unpack_params" class="doc doc-heading">
            <code class="highlight language-python">unpack_params(params, num_psites)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to unpack the parameters for the distributive ODE system.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>params(np.array)</code></b>
              –
              <div class="doc-md-description">
                <p>Parameter vector containing A, B, C, D, S_1.S_n, Ddeg_1.Ddeg_m.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites(int)</code></b>
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>A</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>mRNA production rate.</p>
              </div>
            </li>
            <li>
<b><code>B</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>mRNA degradation rate.</p>
              </div>
            </li>
            <li>
<b><code>C</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>protein production rate.</p>
              </div>
            </li>
            <li>
<b><code>D</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>protein degradation rate.</p>
              </div>
            </li>
            <li>
<b><code>S_rates</code></b>(                  <code><span title="numpy.array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Phosphorylation rates for each site.</p>
              </div>
            </li>
            <li>
<b><code>D_rates</code></b>(                  <code><span title="numpy.array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Dephosphorylation rates for each site.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.distmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python">solve_ode(params, init_cond, num_psites, t)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Solve the ODE system for the distributive phosphorylation model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>params</code></b>
              –
              <div class="doc-md-description">
                <p>array of parameters</p>
              </div>
            </li>
            <li>
              <b><code>init_cond</code></b>
              –
              <div class="doc-md-description">
                <p>initial conditions</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
              –
              <div class="doc-md-description">
                <p>number of phosphorylation sites</p>
              </div>
            </li>
            <li>
              <b><code>t</code></b>
              –
              <div class="doc-md-description">
                <p>time points</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>sol</code></b>              –
              <div class="doc-md-description">
                <p>solution of the ODE system</p>
              </div>
            </li>
            <li>
<b><code>P_fitted</code></b>              –
              <div class="doc-md-description">
                <p>phosphorylated sites</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.randmod" class="doc doc-heading">
            <code>models.randmod</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.randmod.unpack_params" class="doc doc-heading">
            <code class="highlight language-python">unpack_params(params, num_sites)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Unpack parameters for the Random model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>params</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Parameter vector containing A, B, C, D, S_1.S_n, Ddeg_1.Ddeg_m.</p>
              </div>
            </li>
            <li>
              <b><code>num_sites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>A</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>mRNA production rate.</p>
              </div>
            </li>
            <li>
<b><code>B</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>mRNA degradation rate.</p>
              </div>
            </li>
            <li>
<b><code>C</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>protein production rate.</p>
              </div>
            </li>
            <li>
<b><code>D</code></b>(                  <code><span title="float">float</span></code>
)              –
              <div class="doc-md-description">
                <p>protein degradation rate.</p>
              </div>
            </li>
            <li>
<b><code>S</code></b>(                  <code><span title="numpy.array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Phosphorylation rates for each site.</p>
              </div>
            </li>
            <li>
<b><code>Ddeg</code></b>(                  <code><span title="numpy.array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Degradation rates for phosphorylated states.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.randmod.ode_system" class="doc doc-heading">
            <code class="highlight language-python">ode_system(y, t, A, B, C, D, num_sites, S, Ddeg, mono_idx, forward, drop, fcounts, dcounts)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Compute the time derivatives of a random phosphorylation ODE system.</p>
<p>This function supports a large number of phosphorylation states by using
precomputed transition indices to optimize speed.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>y</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Current state vector [R, P, X_1, ..., X_m].</p>
              </div>
            </li>
            <li>
              <b><code>t</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Time (unused; present for compatibility with ODE solvers).</p>
              </div>
            </li>
            <li>
              <b><code>A</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>mRNA production rate.</p>
              </div>
            </li>
            <li>
              <b><code>B</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>mRNA degradation rate.</p>
              </div>
            </li>
            <li>
              <b><code>C</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>protein production rate.</p>
              </div>
            </li>
            <li>
              <b><code>D</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>protein degradation rate.</p>
              </div>
            </li>
            <li>
              <b><code>num_sites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>S</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Phosphorylation rates for each site.</p>
              </div>
            </li>
            <li>
              <b><code>Ddeg</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Degradation rates for phosphorylated states.</p>
              </div>
            </li>
            <li>
              <b><code>mono_idx</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Precomputed indices for mono-phosphorylated states.</p>
              </div>
            </li>
            <li>
              <b><code>forward</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Forward phosphorylation target states.</p>
              </div>
            </li>
            <li>
              <b><code>drop</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Dephosphorylation target states.</p>
              </div>
            </li>
            <li>
              <b><code>fcounts</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of valid forward transitions for each state.</p>
              </div>
            </li>
            <li>
              <b><code>dcounts</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of valid dephosphorylation transitions for each state.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>out</code></b>(                  <code><span title="numpy.array">array</span></code>
)              –
              <div class="doc-md-description">
                <p>Derivatives [dR, dP, dX_1, ..., dX_m].</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.randmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python">solve_ode(popt, y0, num_sites, t)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Integrate the ODE system for phosphorylation dynamics in random phosphorylation model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>popt</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Optimized parameter vector [A, B, C, D, S_1.S_n, Ddeg_1.Ddeg_m].</p>
              </div>
            </li>
            <li>
              <b><code>y0</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Initial condition vector [R0, P0, X1_0, ..., Xm_0].</p>
              </div>
            </li>
            <li>
              <b><code>num_sites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>t</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points to integrate over.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>sol</code></b>(                  <code><span title="ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>Full ODE solution of shape (len(t), len(y0)).</p>
              </div>
            </li>
            <li>
<b><code>mono</code></b>(                  <code><span title="ndarray">ndarray</span></code>
)              –
              <div class="doc-md-description">
                <p>1D array of fitted values for R (after OFFSET) and P states.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="models.succmod" class="doc doc-heading">
            <code>models.succmod</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="models.succmod.ode_core" class="doc doc-heading">
            <code class="highlight language-python">ode_core(y, t, A, B, C, D, S_rates, D_rates)</code>

</h3>


    <div class="doc doc-contents ">

        <p>The core of the ODE system for the successive ODE model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>y</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>The current state of the system.</p>
              </div>
            </li>
            <li>
              <b><code>t</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>The current time.</p>
              </div>
            </li>
            <li>
              <b><code>A</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>The mRNA production rate.</p>
              </div>
            </li>
            <li>
              <b><code>B</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>The mRNA degradation rate.</p>
              </div>
            </li>
            <li>
              <b><code>C</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>The protein production rate.</p>
              </div>
            </li>
            <li>
              <b><code>D</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>The protein degradation rate.</p>
              </div>
            </li>
            <li>
              <b><code>S_rates</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>The phosphorylation rates for each site.</p>
              </div>
            </li>
            <li>
              <b><code>D_rates</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>The dephosphorylation rates for each site.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    dydt (np.array): The derivatives of the state variables.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.succmod.unpack_params" class="doc doc-heading">
            <code class="highlight language-python">unpack_params(params, num_psites)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to unpack the parameters for the ODE system.
The parameters are expected to be in the following order:
A, B, C, D, S_rates, D_rates
where S_rates and D_rates are arrays of length num_psites.
The function returns the unpacked parameters as separate variables.
:param params: array of parameters
:param num_psites: number of phosphorylation sites
:return: A, B, C, D, S_rates, D_rates</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="models.succmod.solve_ode" class="doc doc-heading">
            <code class="highlight language-python">solve_ode(params, init_cond, num_psites, t)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Solve the ODE system using the given parameters and initial conditions.
The function integrates the ODE system over time and returns the solution.</p>
<p>:param params:
:param init_cond:
:param num_psites:
:param t:
:return: solution, solution of phosphorylated sites</p>


    </div>

</div>



  </div>

    </div>

</div><h3 id="steady-state-calculation">Steady-State Calculation</h3>


<div class="doc doc-object doc-module">



<h2 id="steady.initdist" class="doc doc-heading">
            <code>steady.initdist</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="steady.initdist.initial_condition" class="doc doc-heading">
            <code class="highlight language-python">initial_condition(num_psites: int) -&gt; list</code>

</h3>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for distributive phosphorylation model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>list</code></b>(                  <code><span title="list">list</span></code>
)              –
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="steady.initrand" class="doc doc-heading">
            <code>steady.initrand</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="steady.initrand.initial_condition" class="doc doc-heading">
            <code class="highlight language-python">initial_condition(num_psites: int) -&gt; list</code>

</h3>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for random phosphorylation model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>list</code></b>(                  <code><span title="list">list</span></code>
)              –
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="steady.initsucc" class="doc doc-heading">
            <code>steady.initsucc</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="steady.initsucc.initial_condition" class="doc doc-heading">
            <code class="highlight language-python">initial_condition(num_psites: int) -&gt; list</code>

</h3>


    <div class="doc doc-contents ">

        <p>Calculates the initial steady-state conditions for a given number of phosphorylation sites
for successive phosphorylation model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites in the model.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>list</code></b>(                  <code><span title="list">list</span></code>
)              –
              <div class="doc-md-description">
                <p>A list of steady-state values for the variables [R, P, P_sites].</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="ValueError">ValueError</span></code>
              –
              <div class="doc-md-description">
                <p>If the optimization fails to find a solution for the steady-state conditions.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div><h3 id="plotting">Plotting</h3>


<div class="doc doc-object doc-module">



<h2 id="plotting.plotting" class="doc doc-heading">
            <code>plotting.plotting</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="plotting.plotting.Plotter" class="doc doc-heading">
            <code>Plotter</code>


</h3>


    <div class="doc doc-contents ">


        <p>A class to encapsulate plotting functionalities for ODE model analysis.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Attributes:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The gene or experiment name.</p>
              </div>
            </li>
            <li>
              <b><code>out_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>The directory where plots will be saved.</p>
              </div>
            </li>
            <li>
              <b><code>color_palette</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of color codes used for plotting.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_parallel" class="doc doc-heading">
            <code class="highlight language-python">plot_parallel(solution: np.ndarray, labels: list)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots a parallel coordinates plot for the given solution.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>solution</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>2D numpy array of shape (samples, features) representing the data.</p>
              </div>
            </li>
            <li>
              <b><code>labels</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of labels for the features in the solution.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.pca_components" class="doc doc-heading">
            <code class="highlight language-python">pca_components(solution: np.ndarray, target_variance: float = 0.99)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots a scree plot showing the explained variance ratio for PCA components.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>solution</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>2D numpy array of shape (samples, features) representing the data.</p>
              </div>
            </li>
            <li>
              <b><code>target_variance</code></b>
                  (<code><span title="float">float</span></code>, default:
                      <code>0.99</code>
)
              –
              <div class="doc-md-description">
                <p>The target variance to explain. Defaults to 0.99.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_pca" class="doc doc-heading">
            <code class="highlight language-python">plot_pca(solution: np.ndarray, components: int = 3)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the PCA results for the given solution.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>solution</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>2D numpy array of shape (samples, features) representing the data.</p>
              </div>
            </li>
            <li>
              <b><code>components</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>3</code>
)
              –
              <div class="doc-md-description">
                <p>Number of PCA components to plot. Defaults to 3.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>tuple</code></b>              –
              <div class="doc-md-description">
                <p>PCA result and explained variance ratio.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_tsne" class="doc doc-heading">
            <code class="highlight language-python">plot_tsne(solution: np.ndarray, perplexity: int = 30)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots a t-SNE visualization of the given solution.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>solution</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>2D numpy array of shape (samples, features) representing the data.</p>
              </div>
            </li>
            <li>
              <b><code>perplexity</code></b>
                  (<code><span title="int">int</span></code>, default:
                      <code>30</code>
)
              –
              <div class="doc-md-description">
                <p>The perplexity parameter for t-SNE. Defaults to 30.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>np.ndarray: The t-SNE result.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_param_series" class="doc doc-heading">
            <code class="highlight language-python">plot_param_series(estimated_params: list, param_names: list, time_points: np.ndarray)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the time series of estimated parameters over the given time points.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>estimated_params</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of estimated parameters.</p>
              </div>
            </li>
            <li>
              <b><code>param_names</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of parameter names.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of time points.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_profiles" class="doc doc-heading">
            <code class="highlight language-python">plot_profiles(data: pd.DataFrame)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the profiles of estimated parameters over time.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing the time series data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_model_fit" class="doc doc-heading">
            <code class="highlight language-python">plot_model_fit(model_fit: np.ndarray, Pr_data: np.ndarray, P_data: np.ndarray, R_data: np.ndarray, sol: np.ndarray, num_psites: int, psite_labels: list, time_points: np.ndarray)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the model fit for mRNA, protein, and phosphorylated species across time.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>model_fit</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Flattened model fit data (length = 9 + 14 + 14*num_psites).</p>
              </div>
            </li>
            <li>
              <b><code>Pr_data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Protein data (14,).</p>
              </div>
            </li>
            <li>
              <b><code>P_data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Phosphorylation data (num_psites + 2, 14).</p>
              </div>
            </li>
            <li>
              <b><code>R_data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>mRNA data (9,).</p>
              </div>
            </li>
            <li>
              <b><code>sol</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>ODE solution array.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Labels for phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points (14,).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_param_scatter" class="doc doc-heading">
            <code class="highlight language-python">plot_param_scatter(est_arr: np.ndarray, num_psites: int, time_vals: np.ndarray)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots scatter and density plots for parameters.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>est_arr</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>2D numpy array of estimated parameters.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>time_vals</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Array of time values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_heatmap" class="doc doc-heading">
            <code class="highlight language-python">plot_heatmap(param_value_df: pd.DataFrame)</code>

</h4>


    <div class="doc doc-contents ">



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>param_value_df</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing parameter values with 'Protein' as one of the columns.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_error_distribution" class="doc doc-heading">
            <code class="highlight language-python">plot_error_distribution(error_df: pd.DataFrame)</code>

</h4>


    <div class="doc doc-contents ">



<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>error_df</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>DataFrame containing errors with 'MAE' as one of the columns.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_gof" class="doc doc-heading">
            <code class="highlight language-python">plot_gof(merged_data: pd.DataFrame)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plot the goodness of fit for the model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>merged_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>Dataframe containing merged data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_kld" class="doc doc-heading">
            <code class="highlight language-python">plot_kld(merged_data: pd.DataFrame)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the Kullback-Divergence for the model.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>merged_data</code></b>
                  (<code><span title="pandas.DataFrame">DataFrame</span></code>)
              –
              <div class="doc-md-description">
                <p>Dataframe containing merged data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_params_bar" class="doc doc-heading">
            <code class="highlight language-python">plot_params_bar(ci_results: dict, param_labels: list = None)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots bar plot for estimated parameter with 95% Confidence Interval.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>ci_results</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing the results of the confidence intervals.</p>
              </div>
            </li>
            <li>
              <b><code>param_labels</code></b>
                  (<code><span title="list">list</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>List of parameter labels. Defaults to None.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_knockouts" class="doc doc-heading">
            <code class="highlight language-python">plot_knockouts(results_dict: dict, num_psites: int, psite_labels: list)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plot wild-type and knockout simulation results for comparison.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>results_dict</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>Dictionary containing simulation results.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of phosphorylation site labels.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_top_param_pairs" class="doc doc-heading">
            <code class="highlight language-python">plot_top_param_pairs(excel_path: str)</code>

</h4>


    <div class="doc doc-contents ">

        <p>For each gene's '_perturbations' sheet in the Excel file,
plot scatter plots for the parameter pairs with correlation.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>excel_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the Excel file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_model_perturbations" class="doc doc-heading">
            <code class="highlight language-python">plot_model_perturbations(problem: dict, Si: dict, cutoff_idx: int, time_points: np.ndarray, n_sites: int, best_model_psite_solutions: np.ndarray, best_mrna_solutions: np.ndarray, best_protein_solutions: np.ndarray, psite_labels: list[str], protein_data_ref: np.ndarray, psite_data_ref: np.ndarray, rna_ref: np.ndarray, model_fit_sol: np.ndarray) -&gt; None</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plot the best model perturbations for the given data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>problem</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>The optimization problem.</p>
              </div>
            </li>
            <li>
              <b><code>Si</code></b>
                  (<code><span title="dict">dict</span></code>)
              –
              <div class="doc-md-description">
                <p>The simulation index.</p>
              </div>
            </li>
            <li>
              <b><code>cutoff_idx</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>The cutoff index for the time points.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The time points for the data.</p>
              </div>
            </li>
            <li>
              <b><code>n_sites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>The number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>best_model_psite_solutions</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The best model phosphorylation site solutions.</p>
              </div>
            </li>
            <li>
              <b><code>best_mrna_solutions</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The best model mRNA solutions.</p>
              </div>
            </li>
            <li>
              <b><code>best_protein_solutions</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The best model protein solutions.</p>
              </div>
            </li>
            <li>
              <b><code>protein_ref</code></b>
              –
              <div class="doc-md-description">
                <p>The reference data for the protein.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels</code></b>
                  (<code><span title="list">list</span>[<span title="str">str</span>]</code>)
              –
              <div class="doc-md-description">
                <p>The labels for the phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>psite_data_ref</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The reference data for the phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>rna_ref</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The reference data for mRNA.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_time_state_grid" class="doc doc-heading">
            <code class="highlight language-python">plot_time_state_grid(samples: np.ndarray, time_points: np.ndarray, state_names: list)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Grid of strip plots per state showing variability across time.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>samples</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>shape (n_samples, n_timepoints, n_states)</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>array of time points</p>
              </div>
            </li>
            <li>
              <b><code>state_names</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>list of state names</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_phase_space" class="doc doc-heading">
            <code class="highlight language-python">plot_phase_space(samples: np.ndarray, state_names: list)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Phase space plots: one state vs another for each simulation.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>samples</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Shape (n_samples, n_timepoints, n_states)</p>
              </div>
            </li>
            <li>
              <b><code>state_names</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of state names (length = num_states)</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_future_fit" class="doc doc-heading">
            <code class="highlight language-python">plot_future_fit(P_data: np.ndarray, R_data: np.ndarray, sol: np.ndarray, num_psites: int, psite_labels: list, time_points: np.ndarray)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Plots the model fit for the future time points.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>P_data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Data for phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>R_data</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Data for mRNA.</p>
              </div>
            </li>
            <li>
              <b><code>sol</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Model solution.</p>
              </div>
            </li>
            <li>
              <b><code>num_psites</code></b>
                  (<code><span title="int">int</span></code>)
              –
              <div class="doc-md-description">
                <p>Number of phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>psite_labels</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>Labels for phosphorylation sites.</p>
              </div>
            </li>
            <li>
              <b><code>time_points</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>Time points for the data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_regularization" class="doc doc-heading">
            <code class="highlight language-python">plot_regularization(excel_path: str)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Read every '<gene>_params' sheet in the Excel file, pull the Regularization value,
and plot a horizontal bar chart of regularization vs. gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>excel_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the Excel file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="plotting.plotting.Plotter.plot_model_error" class="doc doc-heading">
            <code class="highlight language-python">plot_model_error(excel_path: str)</code>

</h4>


    <div class="doc doc-contents ">

        <p>Read every '<gene>_params' sheet in the Excel file, pull the RMSE value,
and plot a horizontal bar chart of RMSE vs. gene.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>excel_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the Excel file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h3 id="utility-functions">Utility Functions</h3>


<div class="doc doc-object doc-module">



<h2 id="utils.display" class="doc doc-heading">
            <code>utils.display</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="utils.display.ensure_output_directory" class="doc doc-heading">
            <code class="highlight language-python">ensure_output_directory(directory)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Ensure the output directory exists. If it doesn't, create it.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>directory</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the output directory.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.load_data" class="doc doc-heading">
            <code class="highlight language-python">load_data(excel_file, sheet='Estimated Values')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Load data from an Excel file. The default sheet is "Estimated Values".</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>excel_file</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the Excel file.</p>
              </div>
            </li>
            <li>
              <b><code>sheet</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;Estimated Values&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Name of the sheet to load. Default is "Estimated Values".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: DataFrame containing the data from the specified sheet.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.format_duration" class="doc doc-heading">
            <code class="highlight language-python">format_duration(seconds)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Format a duration in seconds into a human-readable string.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>seconds</code></b>
                  (<code><span title="float">float</span></code>)
              –
              <div class="doc-md-description">
                <p>Duration in seconds.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    str: Formatted duration string.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.merge_obs_est" class="doc doc-heading">
            <code class="highlight language-python">merge_obs_est(filename)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to merge observed and estimated data from an Excel file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filename</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the Excel file containing observed and estimated data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: Merged DataFrame containing observed and estimated values for each gene and Psite.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.save_result" class="doc doc-heading">
            <code class="highlight language-python">save_result(results, excel_filename)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Function to save results to an Excel file.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>results</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of dictionaries containing results for each gene.</p>
              </div>
            </li>
            <li>
              <b><code>excel_filename</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the output Excel file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.create_report" class="doc doc-heading">
            <code class="highlight language-python">create_report(results_dir: str, output_file: str = f'{model_type}_report.html')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Creates a single global report HTML file from all gene folders inside the results directory.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>results_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the root result's directory.</p>
              </div>
            </li>
            <li>
              <b><code>output_file</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>f&#39;{<span title="config.constants.model_type">model_type</span>}_report.html&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Name of the generated global report file (placed inside results_dir).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.display.organize_output_files" class="doc doc-heading">
            <code class="highlight language-python">organize_output_files(directories: Iterable[Union[str, Path]])</code>

</h3>


    <div class="doc doc-contents ">

        <p>Organize output files into protein-specific folders and a general folder.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>directories</code></b>
                  (<code><span title="typing.Iterable">Iterable</span>[<span title="typing.Union">Union</span>[<span title="str">str</span>, <span title="pathlib.Path">Path</span>]]</code>)
              –
              <div class="doc-md-description">
                <p>List of directories to organize.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="utils.tables" class="doc doc-heading">
            <code>utils.tables</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="utils.tables.generate_tables" class="doc doc-heading">
            <code class="highlight language-python">generate_tables(xlsx_file_path)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generate hierarchical tables from the XLSX file containing alpha and beta values.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>xlsx_file_path</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the XLSX file containing alpha and beta values.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>tuple</code></b>              –
              <div class="doc-md-description">
                <p>containing protein, psite, and the corresponding table.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.tables.save_tables" class="doc doc-heading">
            <code class="highlight language-python">save_tables(tables, output_dir)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Save the generated tables as LaTeX and CSV files.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>tables</code></b>
                  (<code><span title="list">list</span></code>)
              –
              <div class="doc-md-description">
                <p>List of tuples containing protein, psite, and the corresponding table.</p>
              </div>
            </li>
            <li>
              <b><code>output_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Directory to save the LaTeX and CSV files.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.tables.save_master_table" class="doc doc-heading">
            <code class="highlight language-python">save_master_table(folder='latex', output_file='latex/all_tables.tex')</code>

</h3>


    <div class="doc doc-contents ">

        <p>Save a master LaTeX file that includes all individual LaTeX files from the specified folder.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>folder</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;latex&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The folder containing the individual LaTeX files.</p>
              </div>
            </li>
            <li>
              <b><code>output_file</code></b>
                  (<code><span title="str">str</span></code>, default:
                      <code>&#39;latex/all_tables.tex&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The name of the master LaTeX file to be created.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="utils.latexit" class="doc doc-heading">
            <code>utils.latexit</code>


</h2>

    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="utils.latexit.generate_latex_table" class="doc doc-heading">
            <code class="highlight language-python">generate_latex_table(df, sheet_name)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generate LaTeX code for a table from a DataFrame.
Args:
    df (pd.DataFrame): DataFrame to convert to LaTeX.
    sheet_name (str): Name of the sheet for caption and label.
Returns:
    str: LaTeX code for the table.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.latexit.generate_latex_image" class="doc doc-heading">
            <code class="highlight language-python">generate_latex_image(image_filename)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Generate LaTeX code for an image.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>image_filename</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Path to the image file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>        <p>Returns:
    str: LaTeX code for the image.</p>


    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="utils.latexit.main" class="doc doc-heading">
            <code class="highlight language-python">main(input_dir)</code>

</h3>


    <div class="doc doc-contents ">

        <p>Main function to process Excel and PNG files in the input directory and generate LaTeX code.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>input_dir</code></b>
                  (<code><span title="str">str</span></code>)
              –
              <div class="doc-md-description">
                <p>Directory containing Excel and PNG files.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

    </div>

</div>



  </div>

    </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="index.html" class="btn btn-neutral float-left" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="Documentation/CHANGELOG.html" class="btn btn-neutral float-right" title="Changelog">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Copyright &copy; 2025 <a href="https://bibymaths.github.io">Abhinav Mishra</a>.</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/bibymaths/phoskintime" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="index.html" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="Documentation/CHANGELOG.html" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="js/jquery-3.6.0.min.js"></script>
    <script>var base_url = ".";</script>
    <script src="js/theme_extra.js"></script>
    <script src="js/theme.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/katex.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/katex@0.16.0/dist/contrib/auto-render.min.js"></script>
      <script src="js/katex-setup.js"></script>
      <script src="search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
